---
title: "mapping_qualities"
author: "Rachel Xu"
date: "4/14/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
# library(knitr)
library(tidyverse)
library(reshape2)
library(gridExtra)
library(scales)
library(readr)

polyapath <- "/Users/rx32940/Dropbox/5.Rachel-projects/Transcriptomics_PolyATail/polyAtail"
```
# example code
https://www.biostars.org/p/412621/s
```{r}


samples <- c("Q29_Copenhageni_Basecalled_May_22_2020_Direct-RNA","Q36_Copenhageni_Basecalled_June_9_2020-Repeat_Direct-RNA","Q29_Copenhageni_Basecalled-June_11_2020_Repeat_Direct-RNA", "Q36_Copenhageni_Basecalled_June_9_2020-Repeat_Direct-RNA")

samp_ab <- c("Q29.Cop.DRNA","Q36.Cop.DRNA", "Q29.Cop.DRNA.R","Q36.Cop.DRNA.R")

total.reads<- data.frame(sample=c(), total.reads=c())

# map to genome
mapped.reads.genome <- data.frame(sample=c(), mapped.reads=c())
# map to transcriptome (set minimap -p, min secondary-to-primary score ratio, [0.99])
mapped.reads.trans <- data.frame(sample=c(), mapped.reads.trans=c())

# samtools flagstats 
mapped.second.genome <- data.frame(sample=c(), secondary.alignment=c())
mapped.supp.genome <- data.frame(sample=c(), supplementary.alignment=c())
mapped.second.trans <- data.frame(sample=c(), secondary.alignment.trans=c())
mapped.supp.trans <- data.frame(sample=c(), supplementary.alignment.trans=c())

# total mapped reads/total reads
mapped.percentage <- data.frame(sample=c(), mapped.percentage=c())
mapped.percentage.trans <- data.frame(sample=c(), mapped.percentage.trans=c())


for (index in 1:length(samples)){
  temp_genome_data <- readLines(file.path(polyapath, "genome_sam_stats", paste0(samples[index],".sam.stats")))
  temp_trans_data <- readLines(file.path(polyapath, "trans_sam_stats", paste0(samples[index],".sam.stats")))
  
  temp_sn_genome <- grep("^SN",temp_genome_data, value=TRUE)
  temp_sn_trans <- grep("^SN",temp_trans_data, value=TRUE)
  
  temp_sn_genome <- separate(data.frame(temp_sn_genome),col=1, into=c("ID", "Name","Value"), sep="\t")[,-1]
  temp_sn_trans <- separate(data.frame(temp_sn_trans),col=1, into=c("ID", "Name","Value"), sep="\t")[,-1]
  
  # following number can alse get from samtools view -c $file.bam/sam -F (samtools bit code)
  # check bit code at: http://broadinstitute.github.io/picard/explain-flags.html
  # -F 2304, all primary reads,excluding reads non-primary aligned and supplementary alignment (but not necessarily mapped)
  total.reads <- rbind(total.reads,data.frame(sample=samp_ab[index], total.reads=c(temp_sn_genome[1,]$Value)))
  # -F 2308, all primary reads,excluding reads that are not primarily aligned and supplementary alignment (mapped)
  mapped.reads.genome <- rbind(mapped.reads.genome,data.frame(sample=samp_ab[index], mapped.reads.genome=c(temp_sn_genome[7,]$Value)))
  mapped.reads.trans <- rbind(mapped.reads.trans,data.frame(sample=samp_ab[index], mapped.reads.trans=c(temp_sn_trans[7,]$Value)))
  
  # samtools flagstats results for secondard and supplementary reads stats
  flagstats.genome <- readLines(file.path(polyapath, "genome_flagstats", paste0(samples[index],".flagstats")))
  mapped.second.genome <- rbind(mapped.second.genome,data.frame(sample=samp_ab[index], secondary.alignment=unlist(strsplit(flagstats.genome[2], split=" ", fixed=TRUE))[1]))
  mapped.supp.genome <- rbind(mapped.supp.genome,data.frame(sample=samp_ab[index],supplementary.alignment=unlist(strsplit(flagstats.genome[3], split=" ", fixed=TRUE))[1]))
  
    flagstats.trans <- readLines(file.path(polyapath, "trans_flagstats", paste0(samples[index],".flagstats")))
  mapped.second.trans <- rbind(mapped.second.trans,data.frame(sample=samp_ab[index], secondary.alignment.trans=unlist(strsplit(flagstats.trans[2], split=" ", fixed=TRUE))[1]))
  mapped.supp.trans <- rbind(mapped.supp.trans,data.frame(sample=samp_ab[index],supplementary.alignment.trans=unlist(strsplit(flagstats.trans[3], split=" ", fixed=TRUE))[1]))
  
   mapped.percentage <- rbind(mapped.percentage,data.frame(sample=samp_ab[index],mapped.percentage=parse_number(unlist(strsplit(flagstats.genome[5], split=" ", fixed=TRUE))[5])))
     mapped.percentage.trans <- rbind(mapped.percentage.trans,data.frame(sample=samp_ab[index],mapped.percentage.trans=parse_number(unlist(strsplit(flagstats.trans[5], split=" ", fixed=TRUE))[5])))
}



# sn <- readLines(file.path(polyapath, "genome_sam_stats", paste0("Copenhageni_Basecalled_Aug_16_2019_Direct-cDNA_NoPolyATail",".sam.stats")))
# sn <- grep("^SN",sn, value=TRUE)
# sn <- separate(data.frame(sn),col=1, into=c("ID", "Name","Value"), sep="\t")[,-1]
# head(sn,10)
# kable(sn, caption="Summary numbers")
```


# plot primary mapped reads (total vs. genome vs. transcriptome)
```{r}
map.stats.combined <- merge(merge(total.reads,mapped.reads.genome), mapped.reads.trans)

map.stats.combined
map.stats.longer <- pivot_longer(map.stats.combined, cols = c("total.reads", "mapped.reads.genome", "mapped.reads.trans"), names_to = "primary.alignment.stats", values_to = "Number")
map.stats.longer$Number <- as.numeric(map.stats.longer$Number)
map.stats.longer$primary.alignment.stats <- factor(map.stats.longer$primary.alignment.stats, levels = c("total.reads", "mapped.reads.genome", "mapped.reads.trans"))
map.stats.longer$sample <- factor(map.stats.longer$sample, levels=c("Q29.Cop.DRNA","Q29.Cop.DRNA.R","Q36.Cop.DRNA", "Q36.Cop.DRNA.R"))
# kable(total.reads, caption="Total Number of Reads")

ggplot(map.stats.longer, aes(x= sample, y= Number))+
  geom_bar(stat="identity", position=position_dodge2(padding =0),aes(fill=primary.alignment.stats))+
    theme_bw()+
  theme(axis.text.x = element_text(angle = 90))+
  scale_y_continuous(label=comma)+
  labs(y="Number of Reads")+
  scale_fill_manual(values=c("gray", "orange", "blue"))
  
```

# plot primary mapped reads (total vs. genome vs. transcriptome)
```{r}

map.flag.combined <- merge(merge(merge(mapped.second.genome,mapped.second.trans), merge(mapped.supp.genome,mapped.supp.trans)), map.stats.combined) %>% dplyr::select(!total.reads)
map.flag.longer <- pivot_longer(map.flag.combined, cols = c(secondary.alignment, secondary.alignment.trans, supplementary.alignment, supplementary.alignment.trans), names_to = "Non-primary.alignment.stats", values_to = "Number")
map.flag.longer$Number <- as.numeric(map.flag.longer$Number)

map.flag.longer$`Non-primary.alignment.stats` <- factor(map.flag.longer$`Non-primary.alignment.stats`, levels = c("secondary.alignment", "secondary.alignment.trans", "supplementary.alignment", "supplementary.alignment.trans"))
map.flag.longer
map.flag.longer$Reference <- rep(c("genome", "transcriptome"),8)
map.flag.longer$align_type <- rep(c(rep("secondary",2),rep("supplementary",2)),4)
map.flag.longer

proportion <- ifelse(map.flag.longer$Reference == "genome",map.flag.longer$Number/as.numeric(map.flag.longer$mapped.reads.genome), map.flag.longer$Number/as.numeric(map.flag.longer$mapped.reads.trans))

map.flag.longer$proportion <- proportion

map.flag.longer$sample <- factor(map.flag.longer$sample, levels=c("Q29.Cop.DRNA","Q29.Cop.DRNA.R","Q36.Cop.DRNA", "Q36.Cop.DRNA.R"))



ggplot(map.flag.longer, aes(x= sample, y= proportion))+
  geom_bar(stat="identity", position=position_dodge2(),aes(fill=align_type))+
  facet_wrap(vars(Reference),ncol=1)+
    theme_bw()+
  theme(axis.text.x = element_text(angle = 90))+
  scale_y_continuous(labels = scales::percent_format(accuracy = 1))+
  labs(y="Non-primary Vs. Primary Reads (%)")
  
```
# Read Length
```{r}

read.length <- data.frame(`Read Length`=c(), count=c(), sample=c())
for (index in 1:length(samples)){
  temp_genome_data <- readLines(file.path(polyapath, "genome_sam_stats", paste0(samples[index],".sam.stats")))

  temp_rl_genome <- grep("^RL",temp_genome_data, value=TRUE)
  
   temp_rl_genome <- separate(data.frame(temp_rl_genome),col=1, into=c("ID","Read Length", "count"), sep="\t")[,-1]
   
   temp_rl_genome <- temp_rl_genome %>% mutate(sample=samp_ab[index])
   read.length <- rbind(read.length, temp_rl_genome)
}
temp_rl_genome

library(plotly)
p <- ggplot(read.length %>% subset(as.numeric(`Read Length`) < 5000), aes(x= as.numeric(`Read Length`), y = as.numeric(count)))+
  geom_line(aes(color=sample))+
  theme_bw()+
  labs(x="Read Length", y="Frequency")
p
ggplotly(p)
```

```{r}

gc.content <- data.frame(GC=c(), count=c(), sample=c())

for (index in 1:length(samples)){
  temp_genome_data <- readLines(file.path(polyapath, "genome_sam_stats", paste0(samples[index],".sam.stats")))
  
  gc <- grep("^GCF",temp_genome_data, value=TRUE)
  gc <- separate(data.frame(gc),col=1, into=c("Pair", "GC", "Count"), sep="\t")[,-1]
   
  gc <- gc %>% mutate(sample=samp_ab[index])
  gc.content  <- rbind(gc.content , gc)
}
gc.content %>% group_by(sample) %>% mutate(total=sum(as.numeric(Count)*2))
library(plotly)
p <- ggplot(gc.content , aes(x= as.numeric(GC), y = as.numeric(Count)/sum(as.numeric(Count))))+
  geom_line(aes(color=sample))+
  theme_bw()+
  labs(x="GC content", y="Count")

ggplotly(p)
```

