---
title: "minION_polyA"
author: "Rachel Xu"
date: "4/28/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

### set path and sample names
```{r}

nano.path<- "/Users/rx32940/Dropbox/5.Rachel-projects/Transcriptomics_PolyATail/polyAtail/nanopolish_output"

man.path <- "/Users/rx32940/Dropbox/5.Rachel-projects/Transcriptomics_PolyATail/manuscript/characterize"

```

# NanoPolish

### Read Nanopolish Files 
```{r}
library(dplyr)
headers <- c("readname","contig","position","leader_start","adapter_start","polya_start","transcript_start","read_rate","polya_length","qc_tag","biotype")

Q29.Cop.DRNA <- read.csv(file.path(nano.path, "Q29_Copenhageni_Direct-RNA_FAST5_Pass.nanopolish.polyA.pass_only.biotypeAnnotated.tsv"),sep = "\t", col.names =headers)
Q29.Cop.DRNA$sample <- "Q29.Cop.DRNA"
Q29.Cop.DRNA$temp <- "Q29"


Q29.Cop.DRNA.R <- read.csv(file.path(nano.path, "Q29_Copenhageni_Direct-RNA_Repeat_FAST5_Pass.nanopolish.polyA.pass_only.biotypeAnnotated.tsv"),sep = "\t", col.names =headers)
Q29.Cop.DRNA.R$sample <- "Q29.Cop.DRNA.R"
Q29.Cop.DRNA.R$temp <- "Q29"


Q36.Cop.DRNA <- read.csv(file.path(nano.path, "Q36_Copenhageni_Direct-RNA_FAST5_Pass.nanopolish.polyA.pass_only.biotypeAnnotated.tsv"),sep = "\t", col.names =headers)
Q36.Cop.DRNA$sample <- "Q36.Cop.DRNA"
Q36.Cop.DRNA$temp <- "Q36"


Q36.Cop.DRNA.R <- read.csv(file.path(nano.path, "Q36_Copenhageni_Direct-RNA_Repeat_FAST5_Pass.nanopolish.polyA.pass_only.biotypeAnnotated.tsv"),sep = "\t", col.names =headers)
Q36.Cop.DRNA.R$sample <- "Q36.Cop.DRNA.R"
Q36.Cop.DRNA.R$temp <- "Q36"

directRna.nano.combined <- rbind(Q29.Cop.DRNA, Q29.Cop.DRNA.R,Q36.Cop.DRNA,Q36.Cop.DRNA.R)
summary(directRna.nano.combined$polya_length)
summary(directRna.nano.combined[directRna.nano.combined$temp=="Q29",]$polya_length)
summary(directRna.nano.combined[directRna.nano.combined$temp=="Q36",]$polya_length)
```


### plot over length of the polya tail for each sample
```{r}
library(ggplot2)
ggplot(directRna.nano.combined, aes(x= sample, y=polya_length))+
  geom_violin(aes(fill=temp))+
  geom_boxplot(width=0.1, outlier.size=1)+
  theme_bw()+
  scale_fill_manual(values=c("#bfc0c0", "#ef8354"))+
  scale_y_continuous(limits = c(0,950), breaks=seq(0,950,100))+
  labs(y="PolyA tail length (bp)")

ggsave(file.path(man.path, "II.10.drna.tail.nanopolish.boxviolin.pdf"), height = 7)
```

# polyA tail length vs. position
```{r}
library(scales)
plot <- ggplot(directRna.nano.combined, aes(x=position, y=polya_length, group=temp))+
  geom_point(aes(color=temp))+
  facet_wrap(vars(sample),ncol = 1)+
  theme_bw()+
  theme(axis.text.x = element_text(angle=90))+
  scale_x_continuous(breaks  = seq(0, 4300000, by = 100000), label=comma)+
  scale_color_manual(values=c("#79aca9", "#ef8354"))

plot
```

# polyA tail abundance vs. position
```{r}
library(ggpmisc)
p <- ggplot(directRna.nano.combined, aes(x=position))+
  #  geom_histogram(bins = 500, aes(fill=temp))+
  # geom_freqpoly(binwidth = 500, aes(color=temp))+
  geom_line(aes(color=temp),stat  = "count")+
  facet_wrap(vars(sample),ncol =1)+
  theme_bw()+
  theme(axis.text.x = element_text(angle=90))+
  scale_color_manual(values=c("#79aca9", "#ef8354"))+
  scale_x_continuous(breaks  = seq(0, 4300000, by = 100000), label=comma)+
  scale_y_continuous(label=comma)+
  labs(y="polyA tail abundance")

pb <- ggplot_build(p)

# get global maximum for each sample
global_maxima<- pb[['data']][[1]] %>% group_by(PANEL) %>% filter(count == max(count)) 
global_maxima$sample <-c("Q29.Cop.DRNA","Q29.Cop.DRNA.R","Q36.Cop.DRNA","Q36.Cop.DRNA.R")

maximum_stat_added <- left_join(directRna.nano.combined,global_maxima)

p <- ggplot(maximum_stat_added, aes(x=position))+
  geom_line(aes(color=temp),stat  = "count")+
  geom_text(aes(label=x))
  facet_wrap(vars(sample),ncol =1)+
  theme_bw()+
  theme(axis.text.x = element_text(angle=90))+
  scale_color_manual(values=c("#79aca9", "#ef8354"))+
  scale_x_continuous(breaks  = seq(0, 4300000, by = 100000), label=comma)+
  scale_y_continuous(label=comma)+
  labs(y="polyA tail abundance")
p

# ggsave("/Users/rx32940/Dropbox/5.Rachel-projects/Transcriptomics_PolyATail/manuscript/figures/dRNA.polyATail_Abundance_vs_position.pdf")
# plot
# pg <- ggplot_build(plot)
# data.frame(pg$data[[1]]) %>% subset(colour=="#79aca9") %>% subset(y==max(y))

```
### format gtf file (gene annotation from gtf file)
- formatted gtf file were ran in the python code ```match_biotype_nanopolish.py``` to determine the biotype each read were mapped to
```{r}
# user <- "rachel"
# ref <- paste0(paste0("/Users/",user),"/Dropbox/5.Rachel-projects/Transcriptomics_PolyATail/quant_results/reference/")
# gtf_df <- read.csv(paste0(ref,"GCF_000007685.1_ASM768v1_genomic.gtf"),skip=5,sep=c("\t"),header = FALSE) %>% subset(V3 == "gene")
# head(gtf_df)
# tx2gene <- gtf_df %>% mutate(salmon_tax = sapply(V9, function(x){
#     v9=trimws(unlist(strsplit(x,";")))
#     id=unlist(strsplit(v9[1]," ",fixed=TRUE))[2]
#     id
# })) %>% mutate(biotype=sapply(V9, function(x){
#     v9=trimws(unlist(strsplit(x,";")))
#     biotype_index = grep("gene_biotype", v9)
#     current=unlist(strsplit(v9[biotype_index]," ",fixed=TRUE))
#     biotype=ifelse(current[1]=="gene_biotype",current[2],NA)
#     biotype
# })) %>% dplyr::select(c("salmon_tax","biotype", V4, V5)) %>% rename(start=V4, end=V5)

# write.csv(tx2gene, "/Users/rachel/Dropbox/5.Rachel-projects/Transcriptomics_PolyATail/polyAtail/biotype_position.csv", quote = FALSE, row.names = FALSE)
```

### plot biotype vs. polyA length
```{r}
library(grid)
library(scales)
unique(directRna.nano.combined$biotype)
p1 <- ggplot(directRna.nano.combined, aes(x=biotype, y=log(polya_length)))+
  geom_violin()+
  geom_boxplot(width=0.1)+
  coord_flip()+
  theme_bw()+
  theme(axis.text.y = element_blank(),axis.title.y =element_blank())+
  scale_x_discrete(position = "top")+
  scale_y_continuous(label=comma) +
  labs(y="poly A tail length bps (log)")

p2 <- ggplot(directRna.nano.combined, aes(x=biotype)) +
  geom_bar(stat="count",  alpha=0.2)+
  geom_text(aes(label = ..count..,y= ..count..), stat="count",vjust =-0.5,size=3)+
  coord_flip()+
  theme_bw()+
  theme(axis.title.y =element_blank(),axis.text.y = element_text(hjust=0.5),legend.position = "none")+
  scale_y_continuous(label=comma)


plot <- grid.arrange(cbind(ggplotGrob(p1), ggplotGrob(p2), size = "last"))
plot
ggsave(file.path(man.path, "II.11.drna.nanopolish.biotype.vs.taillength.pdf"),plot, height = 7)
```

# MAP to Transcriptome 
```{r}
library(dplyr)
library(ggplot2)
direct.path <- "/Users/rx32940/Dropbox/5.Rachel-projects/Transcriptomics_PolyATail/directRNA_trans_09012021"
nano.path <- "/Users/rx32940/Dropbox/5.Rachel-projects/Transcriptomics_PolyATail/directRNA_trans_09012021/nanopolish"
ref <- "/Users/rx32940/Dropbox/5.Rachel-projects/Transcriptomics_PolyATail/directRNA_trans_09012021/reference"

transcript.annot <- read.csv(file.path(ref, "GCF_000007685.1_ASM768v1_feature_table.txt"), sep="\t") %>% subset(X..feature == "CDS") %>% select(-c("X..feature", "class", "assembly", "assembly_unit", "seq_type", "related_accession", "GeneID", "locus_tag","attributes"))
```


```{r}

Q29.trans.pass <- read.csv(file.path(nano.path, "Q29_Copenhageni_Direct-RNA_FAST5_Pass.nanopolish.trans.polyA.tsv"), sep="\t") %>% subset(qc_tag == "PASS")

Q29.trans.pass <- Q29.trans.pass %>% mutate(non.redundant_refseq = sapply(contig, function(x){
  s1 <- unlist(strsplit(x, "_", fixed =TRUE))
  paste(s1[4], s1[5],sep="_")
}))

Q29.trans.pass.anot <- left_join(Q29.trans.pass, transcript.annot)
Q29.trans.pass.anot

Q29.trans.pass.anot.Mapped.stat <- Q29.trans.pass.anot %>% group_by(product_accession, name, symbol, chromosome) %>% summarise(Num.Reads.Mapped = n()) %>% arrange(desc(Num.Reads.Mapped))

# transcripts with longest polyA tailed reads mapped to them
longest.tail.transcripts <- head(Q29.trans.pass.anot %>% arrange(desc(polya_length) ) %>% select(product_accession),10)[,1]
Q29.trans.pass.anot %>% subset(product_accession %in%longest.tail.transcripts)

ggplot(Q29.trans.pass.anot %>% subset(product_accession %in%longest.tail.transcripts), aes(x=non.redundant_refseq,y=polya_length ))+
  geom_violin()+
  geom_boxplot(width = 0.3) +
  scale_y_continuous(breaks = seq(0, 600, 30), 
                   limits = c(0,600), 
                   expand = c(0,0))
# ggsave(file.path(nano.path, "transcript_polyATail.pdf"))

# transcripts with largest median length for all polyAed reads mapped to them
longest_polyA_median <- Q29.trans.pass.anot %>% group_by(product_accession, name, symbol, chromosome) %>% mutate(num_reads_mapped = n())%>% mutate(median_polyA_length = median(polya_length)) %>% arrange(desc(median_polyA_length)) %>% mutate(plot_label = sapply(non.redundant_refseq, function(x){
  paste(x, paste0("n=",n()),sep="\n")
}))  %>% subset(num_reads_mapped >10)

top_10_median <- head(unique(longest_polyA_median$median_polyA_length), 10)
ggplot(longest_polyA_median %>% subset(median_polyA_length %in% top_10_median), aes(x=plot_label,y=polya_length, label = median_polyA_length))+
  geom_violin()+
  geom_boxplot(width = 0.3) +
  scale_y_continuous(breaks = seq(0, 200, 10), 
                   limits = c(0,200), 
                   expand = c(0,0))+
  theme(axis.text.x = element_text(angle=90))
# write.csv(Q29.trans.pass.anot.Mapped.stat, file.path(direct.path, "Q29.trans.pass.anot.MappedTramscript.stat.csv"), row.names = FALSE, quote=FALSE)
```

