---
title: "salmon_quan_diffExp"
author: "Rachel Xu"
date: "4/8/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "")
library(tximeta)
library(tximport)
library(ggplot2)
library(dplyr)
library(tidyr)
library(readr)
library(IsoformSwitchAnalyzeR)
```

# list samples, read int metadata
```{r}
dir <- "/Users/rx32940/Dropbox/5.Rachel-projects/Transcriptomics_PolyATail/quant_results"
salmon_dir <- "/Users/rx32940/Dropbox/5.Rachel-projects/Transcriptomics_PolyATail/quant_results/salmon/"
ref <- "/Users/rx32940/Dropbox/5.Rachel-projects/Transcriptomics_PolyATail/quant_results/reference"
# all samples
salmon_samples<- c(
"Q29_Copenhageni_Basecalled-June_11_2020_Repeat_Direct-RNA",
"Q29_Copenhageni_Basecalled_May_22_2020_Direct-RNA",
"Q36_Copenhageni_Basecalled_June_9_2020-Repeat_Direct-RNA",
"Q36_Copenhageni_Basecalled_May_31_2020_Direct-RNA")


all_meta <- read.csv(paste(dir, "all_flair.txt", sep="/"), sep="\t")
direct_rna_meta <- read.csv(paste(dir, "direct_flair.txt", sep="/"), sep="\t")

```

```{r}
# list all files
salmonfiles <- paste0(salmon_dir, "align_quant/","directRNA/")
salmonQuant <- importIsoformExpression(
    parentDir = as.vector(salmonfiles),
    ignoreAfterBar = FALSE,
)
myDesign <- direct_rna_meta %>% mutate(sampleID = paste0(full_name,"_quant")) %>% dplyr::select(-c(X, full_name,batch)) 
myDesign$condition <- factor(myDesign$condition)
gff <- file.path(ref,"GCF_000007685.1_ASM768v1_genomic.gtf")
fasta <- file.path(ref,"GCF_000007685.1_ASM768v1_transcriptome.gffread.fasta")
aSwitchList <- importRdata(
    isoformCountMatrix   = salmonQuant$counts,
    isoformRepExpression = salmonQuant$abundance,
    designMatrix         = myDesign,
    isoformExonAnnoation = gff,
    isoformNtFasta       = fasta,
    ignoreAfterBar = FALSE,
    showProgress = FALSE
)
salmonQuant$counts
myDesign
```

# read in salmon files
```{r}

# list all files
salmonfiles <- file.path(salmon_dir, "align_quant", "directRNA",paste0(salmon_samples,"_quant"), "quant.sf")
# check if all these files exist
all(file.exists(salmonfiles))
names(salmonfiles) <- salmon_samples
# salmonfiles

# add file path to metadata
direct_rna_metadata <- cbind(direct_rna_meta, files=salmonfiles, stringsAsFactors=FALSE)
direct_rna_metadata$condition <- factor(direct_rna_metadata$condition)

## Import quantifications on the transcript level
txi <- tximport(salmonfiles, type="salmon", txOut=TRUE,
                countsFromAbundance="scaledTPM")
# txi
cts <- txi$counts
# cts[1:3,1:3]
cts <- cts[rowSums(cts) > 0,]
# cts
```


### reference database
```{r}

library(GenomicFeatures)
gff <- file.path(ref,"GCF_000007685.1_ASM768v1_genomic.gff")
txdb.filename <- "GCF_000007685.1_ASM768v1_genomic.sqlite"
txdb <- makeTxDbFromGFF(gff,format="gff")
saveDb(txdb, txdb.filename)

txdb <- loadDb(txdb.filename)
txdf <- select(txdb, keys(txdb, "GENEID"), "TXNAME", "GENEID")
tab <- table(txdf$GENEID)
txdf$ntx <- tab[match(txdf$GENEID, names(tab))]

head(txdf,2)


```


## DRIMSeq
```{r}
txdf

rownames(cts) <- cts_rowname
cts
txdf <- txdf[match(cts_rowname,txdf$GENEID),]
all(rownames(cts) == txdf$GENEID)

rownames(cts) == txdf$GENEID

counts <- data.frame(gene_id=txdf$GENEID, feature_id=txdf$GENEID, cts)
counts <- counts[!is.na(counts$feature_id),]
```
### DRIMSeq
```{r}
samps <- direct_rna_metadata %>% dplyr::rename(sample_id = full_name)
colnames(counts) <- c("gene_id","feature_id",samps$sample_id)
dmDSdata(counts = counts, samples = samps)

```



