---
title: "salmon_quan_diffExp"
author: "Rachel Xu"
date: "4/8/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "")

library(tximport)
library(ggplot2)
library(dplyr)
library(tidyr)

```

# list samples, read int metadata
```{r}
user <- "rachel"


dir <- paste0(paste0("/Users/",user),"/Dropbox/5.Rachel-projects/Transcriptomics_PolyATail/quant_results/")
salmon_dir <- paste0(paste0("/Users/",user),"/Dropbox/5.Rachel-projects/Transcriptomics_PolyATail/quant_results/salmon")
ref <- paste0(paste0("/Users/",user),"/Dropbox/5.Rachel-projects/Transcriptomics_PolyATail/quant_results/reference/")
# all samples
salmon_samples<- c(
"Q29_Copenhageni_Basecalled-June_11_2020_Repeat_Direct-RNA",
"Q29_Copenhageni_Basecalled_May_22_2020_Direct-RNA",
"Q36_Copenhageni_Basecalled_June_9_2020-Repeat_Direct-RNA",
"Q36_Copenhageni_Basecalled_May_31_2020_Direct-RNA")


all_meta <- read.csv(paste(dir, "all_flair.txt", sep="/"), sep="\t")
direct_rna_meta <- read.csv(paste(dir, "direct_flair.txt", sep="/"), sep="\t")

```

# read eggnog-mapper annotation (gene annotation from egg nog mapper)
```{r}

egg_nog <- read.csv(paste0(ref, "GCF_000007685.gffread.eggnogmapper.tsv"),sep="\t",skip=4,header=FALSE)
egg_nog
egg_nog_genename <- egg_nog %>% dplyr::select(c(V1,V6))
colnames(egg_nog_genename) <- c("salmon_tax", "gene_name")
head(egg_nog_genename,2)
```
# format gtf file (gene annotation from gtf file)
```{r}
gtf_df <- read.csv(paste0(ref,"GCF_000007685.1_ASM768v1_genomic.gtf"),skip=5,sep=c("\t"),header = FALSE) %>% subset(V3 == "gene")
# head(gtf_df)

tx2gene <- gtf_df %>% mutate(salmon_tax = sapply(V9, function(x){
    v9=trimws(unlist(strsplit(x,";")))
    id=unlist(strsplit(v9[1]," ",fixed=TRUE))[2]
    id
})) %>% mutate(gene_name=sapply(V9, function(x){
    v9=trimws(unlist(strsplit(x,";")))
    id=unlist(strsplit(v9[1]," ",fixed=TRUE))[2]
    current=unlist(strsplit(v9[4]," ",fixed=TRUE))
    gene_name=ifelse(current[1]=="gene",current[2],NA)
    gene_name
})) %>% dplyr::select(c("salmon_tax","gene_name"))

head(tx2gene,2)
```


# read in salmon files
```{r}

# list all files
salmonfiles <- file.path(salmon_dir, "align_quant", "directRNA",paste0(salmon_samples,"_quant"), "quant.sf")
# check if all these files exist
all(file.exists(salmonfiles))
names(salmonfiles) <- salmon_samples
# salmonfiles

# add file path to metadata
direct_rna_metadata <- cbind(direct_rna_meta, files=salmonfiles, stringsAsFactors=FALSE)
direct_rna_metadata$condition <- factor(direct_rna_metadata$condition)

## Import quantifications on the transcript level
txi <- tximport(salmonfiles, type="salmon", txOut=TRUE,tx2gene = tx2gene)
# txi
cts <- txi$counts
# cts[1:3,1:3]
cts <- cts[rowSums(cts) > 0,]
rownames(cts) <- sapply(as.character(rownames(cts)),function(x){unlist(strsplit(x,"-",fixed=TRUE))[2]})
head(cts,2)
```

```{r}
direct_rna_meta$samples<- c("Q29.2","Q29.1","Q36.2","Q36.1")
salmon_direct_meta <- direct_rna_meta[,c("samples","condition")]
library("DESeq2")
ddsTxi <- DESeqDataSetFromTximport(txi,
                                   colData = salmon_direct_meta,
                                  samples ~ condition)

ddsTxi$condition <- relevel(ddsTxi$condition, ref="Q29")

salmon.deseq <- DESeq(ddsTxi)

deseq.results <- DESeq2::results(salmon.deseq)
deseq.results<-deseq.results[order(deseq.results$padj),]

sum(deseq.results$padj < 0.05 , na.rm=TRUE)
head(deseq.results,2)
```

# Plot PCA to see the clustering of the samples (similarities and variances within a group)
```{r}
ddsTxi
rlg <- rlog(ddsTxi,blind = TRUE)
pcaData <- plotPCA(rlg,ntop=3736,returnData=TRUE)
pcaData$Repeat <- c(rep(c("R2", "R1"),2))
pcaData
percentVar <- round(100 * attr(pcaData, "percentVar"))

ggplot(pcaData, aes(x = PC1, y = PC2, color=condition, shape=Repeat)) +
  geom_point(size =3) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  theme_bw()+
  coord_fixed() +
  ggtitle("PCA with Regularised Log Transformation(rlog)")
```


```{r}

resultsNames(salmon.deseq)
?resultsNames

res_lfcShrink_salmon <- lfcShrink(salmon.deseq,coef = 2, type="apeglm")
sum(res_lfcShrink_salmon$padj < 0.05 , na.rm=TRUE)

```

```{r}
# read DESeq results in to a data frame
deseq_salmonDF <- as.data.frame(res_lfcShrink_salmon)
# write.csv(res_lfcShrink_salmon,paste0(dir,"deseq_salmonDF.csv"))
# add a column for differential expression based on p value and log fold change
deseq_salmonDF$diffexpressed <- "Not differentially expressed"
deseq_salmonDF
# if with p-value < 0.05 and log2fold change > abs(1), label either up or down regulated 
deseq_salmonDF$diffexpressed[deseq_salmonDF$log2FoldChange >= 1 & deseq_salmonDF$padj < 0.05] <- "Upregulated"
deseq_salmonDF$diffexpressed[deseq_salmonDF$log2FoldChange <= -1 & deseq_salmonDF$padj < 0.05] <- "Downregulated"
deseq_salmonDF
# factor so legend is ordered
deseq_salmonDF$diffexpressed <- factor(deseq_salmonDF$diffexpressed, levels=c("Upregulated", "Downregulated", "Not differentially expressed"))
deseq_salmonDF

# all transcript with associated gene names
deseq_salmonDF$salmon_tax <- rownames(deseq_salmonDF)
deseq_salmonDF
deseq_salmonDF <- left_join(deseq_salmonDF,egg_nog_genename,by="salmon_tax")
deseq_salmonDF$delabel <- NA
deseq_salmonDF$delabel[deseq_salmonDF$diffexpressed != "Not differentially expressed"] <- deseq_salmonDF$gene_name[deseq_salmonDF$diffexpressed != "Not differentially expressed"]
deseq_salmonDF
deseq_salmonDF[!is.na(deseq_salmonDF$delabel),]

library(ggrepel)
deseq_salmonDF
ggplot(deseq_salmonDF, aes(x=log2FoldChange,y = -log10(padj)))+
  geom_point(aes(color =diffexpressed))+
  theme_bw()+
  geom_vline(xintercept=c(-1,1), col="gray",linetype="dashed")+
  geom_hline(yintercept=-log10(0.05), col="gray",linetype="dashed")+
  scale_color_manual(values=c("orange", "blue", "black"))+
  geom_text_repel(aes(label=delabel))+
  labs(x="logs(36C/29C)")+
  theme(legend.position = "none")
```

```{r}

# plotMA(res_lfcShrink)

library(ggplot2)
library(scales) # needed for oob parameter
library(viridis)

# Coerce to a data frame
deseq2ResDF <- as.data.frame(res_lfcShrink_salmon)

# Examine this data frame
head(deseq2ResDF)

# Set a boolean column for significance
deseq2ResDF$significant <- ifelse(deseq2ResDF$padj < .05, "Significant",NA)
deseq2ResDF$salmon_tax <- rownames(deseq2ResDF)
deseq2ResDF <- left_join(deseq2ResDF, egg_nog_genename,by="salmon_tax")
deseq2ResDF$delabel <- NA
deseq2ResDF$delabel[!is.na(deseq2ResDF$significant)] <- deseq2ResDF$gene_name[!is.na(deseq2ResDF$significant)]
deseq2ResDF$delabel[!is.na(deseq2ResDF$significant)]
# Plot the results similar to DEseq2
ggplot(deseq2ResDF, aes(log2(baseMean), log2FoldChange, colour=significant)) + 
  geom_point(size=1) + 
  geom_hline(yintercept = 0, colour="gray",size=1.5) + 
  labs(x="mean of normalized counts", y="log fold change")+
    geom_text_repel(aes(label=delabel),color="black")+
  theme_bw()


```
### plot gene counts for some differentially expressed genes
```{r}
library(grid)
deseq2ResDF[order(deseq2ResDF$padj),] %>% subset(gene_name == "rlpA")
deseq2ResDF#%>% subset(gene_name == "dnaA")
plotCounts(ddsTxi, "gene-LIC_RS11290",main="hsp18(gene-LIC_RS11290)-Upregulated" )
plotCounts(ddsTxi,"gene-LIC_RS02730", main="dnaK(gene-LIC_RS02730)-Upregulated")
plotCounts(ddsTxi,"gene-LIC_RS10300", main="clpB(gene-LIC_RS10300)-Upregulated")
plotCounts(ddsTxi, "gene-LIC_RS19270",main="ltg(gene-LIC_RS19270)-downregulated" )
plotCounts(ddsTxi,"gene-LIC_RS23235", main="gene-LIC_RS23235-downregulated")
plotCounts(ddsTxi,"gene-LIC_RS00275", main="rlpA(gene-LIC_RS00275)-downregulated")
plotCounts(ddsTxi, "gene-LIC_RS00005",main="dnaA(gene-LIC_RS00005)" )
plotCounts(ddsTxi,"gene-LIC_RS00015", main="recF(gene-LIC_RS00015)")
plotCounts(ddsTxi,"gene-LIC_RS00025", main="gyrB(gene-LIC_RS00025)")

```

```{r}
library(goseq)
rownames(egg_nog) <- egg_nog$V1
egg_nog_GO <- egg_nog %>% select(c(V7))
head(egg_nog_GO)
go_dict <- setNames(split(egg_nog_GO, seq(nrow(egg_nog_GO))), rownames(egg_nog_GO)) 

head(go_dict)
lapply(go_dict, function(x){
  print(x)
  # strsplit(x, ",",fixed=TRUE)
})
?goseq
```

### heatmap
```{r}
library(pheatmap)
# transforms the count data to the log2 scale in a way which minimizes differences between samples for rows with small counts, and which normalizes with respect to library size.
rld <- rlog(salmon.deseq, blind = FALSE)

# assay is a summarizedExperiment object, matrix like
topVarGenes <- head(order(rowVars(assay(rld)), decreasing=T),50)# top 100 genes' variances 
matrix <- assay(rld)[ topVarGenes, ] # extract genes with top variation aceoss samples
# which(rownames(matrix) == "gene-LIC_RS02730")
matrix <- matrix - rowMeans(matrix)

colnames(matrix)<- c("Q29.Cop.DRNA.R","Q29.Cop.DRNA", "Q36.Cop.DRNA.R","Q36.Cop.DRNA")

gene_id <- as.data.frame(as.character(rownames(matrix)))
colnames(gene_id) <- "salmon_tax"
gene_id <- left_join(gene_id,egg_nog_genename, by="salmon_tax")
gene_id <- gene_id %>% dplyr::mutate(gene_visual=mapply(function(x,y){ifelse(is.na(x)| x=="", y, paste(y, x,sep="|"))}, gene_name, salmon_tax)) 
rownames(matrix) <- gene_id$gene_visual
# write.csv(matrix, "/Users/rachel/Dropbox/5.Rachel-projects/Transcriptomics_PolyATail/directRNA.Samples.Results/plot.info/heatmap/heatmap.values.csv", quote = FALSE)

# select the 'contrast' you want
annotation_data <- data.frame("samples"=c(rep("Q29",2), rep("Q36",2)))
rownames(annotation_data) <- colnames(matrix)
annoCol <- list(category = c("Q29","Q36"))
head(matrix)
pheatmap(matrix, annotation_col = annotation_data, annotation_colors =annoCol, angle_col=45,fontsize_row=5)
```

### EdgeR

### results obtained using prepDE.py3
```{r}
library(edgeR)
counts <- txi$counts
group <- factor(c("Q29","Q29","Q36","Q36"),levels=c("Q29","Q36"))
y <- DGEList(counts=counts,group=group)
y$samples$group
y$samples$group<- relevel(y$samples$group, ref="Q29") # make Q29 samples the base (control)
keep <- filterByExpr(y) # keeps rows that have worthwhile counts in a minumum numberof samples
y <- y[keep,,keep.lib.sizes=FALSE]
y <- calcNormFactors(y)
design <- model.matrix(~group)
y <- estimateDisp(y,design)
fit <- glmQLFit(y,design)
qlf <- glmQLFTest(fit,coef=2)
topTags(qlf)

summary(qlf)
summary(decideTests(qlf))


qlf$table$FDR <- p.adjust(qlf$table$PValue, method="BH")
qlf$table[qlf$table$FDR <0.05,]
head(qlf$table)
```

```{r}

# read DESeq results in to a data frame
edgeR_salmonDF <- as.data.frame(qlf)
# write.csv(res_lfcShrink_salmon,paste0(dir,"deseq_salmonDF.csv"))
# add a column for differential expression based on p value and log fold change
edgeR_salmonDF$diffexpressed <- "Not differentially expressed"
qlf
# if with p-value < 0.05 and log2fold change > abs(1), label either up or down regulated 
edgeR_salmonDF$diffexpressed[edgeR_salmonDF$logFC >= 1 & edgeR_salmonDF$FDR < 0.05] <- "Upregulated"
edgeR_salmonDF$diffexpressed[edgeR_salmonDF$logFC <= -1 & edgeR_salmonDF$FDR < 0.05] <- "Downregulated"


# edgeR_salmonDFOrdered <- edgeR_salmonDF[order(edgeR_salmonDF$FDR),]
# rownames(edgeR_salmonDFOrdered) <- edgeR_salmonDFOrdered$salmon_tax
# edgeR_salmonDFOrdered <- edgeR_salmonDFOrdered %>% dplyr::select(-c(salmon_tax))
# # write.csv(edgeR_salmonDFOrdered, "/Users/rachel/Dropbox/5.Rachel-projects/Transcriptomics_PolyATail/directRNA.Samples.Results/Diff.Exp.Genes/diff.exp.genes.stats.csv")


# factor so legend is ordered
edgeR_salmonDF$diffexpressed <- factor(edgeR_salmonDF$diffexpressed, levels=c("Upregulated", "Downregulated", "Not differentially expressed"))
Q29_upregulated_gene <- edgeR_salmonDF %>% subset(diffexpressed == "Upregulated")

# all transcript with associated gene names
edgeR_salmonDF$salmon_tax <- rownames(edgeR_salmonDF)
edgeR_salmonDF
edgeR_salmonDF <- left_join(edgeR_salmonDF,egg_nog_genename,by="salmon_tax")
edgeR_salmonDF$delabel <- NA
edgeR_salmonDF$delabel[edgeR_salmonDF$diffexpressed != "Not differentially expressed"] <- edgeR_salmonDF$gene_name[edgeR_salmonDF$diffexpressed != "Not differentially expressed"]
edgeR_salmonDF
edgeR_salmonDF[!is.na(edgeR_salmonDF$delabel),]
library(ggrepel)
ggplot(edgeR_salmonDF, aes(x=logFC,y = -log10(FDR)))+
  geom_point(aes(color =diffexpressed))+
  theme_bw()+
  geom_vline(xintercept=c(-1,1), col="gray",linetype="dashed")+
  geom_hline(yintercept=-log10(0.05), col="gray",linetype="dashed")+
  scale_color_manual(values=c("orange", "blue", "black"))+
  geom_text_repel(aes(label=delabel))+
  labs(x="log2(36C/29C)")+
  theme(legend.position = "none")

```
### plot gene counts for some differentially expressed genes
```{r}
y
count.df <- data.frame(y$counts)

count.df$salmon_tax <- rownames(count.df)
count.df.annot <- left_join(edgeR_salmonDF%>%dplyr::select(salmon_tax,gene_name),count.df, by="salmon_tax")
edgeR_salmonDF

upreg.count <- count.df.annot%>% subset(gene_name == "hsp18" | gene_name == "clpB" | gene_name == "dnaK"| gene_name == "ltg" | gene_name == "rlpA"| gene_name == "carF"| gene_name == "dnaA" | gene_name == "recF"| gene_name == "gyrB")
upreg.count <- upreg.count %>%
  mutate(gene_name =  factor(gene_name, levels = c("hsp18","clpB","dnaK","ltg","rlpA","carF","dnaA","recF","gyrB"))) %>%
  arrange(gene_name) 
upreg.count$expression <- c(rep("upregulated",3), rep("downregulated",3),rep("Not Significant ",3) )
upreg.count_longer <- upreg.count %>% pivot_longer(colnames(upreg.count)[-c(1,2,length(upreg.count))], names_to = "sample_name", values_to="count")

upreg.count_longer$temperature <- rep(c(rep("Q29",2), rep("Q36",2)),9)
upreg.count_longer$replicate <- rep(rep(c("R2","R1"),2),9)

upreg.count_longer

ggplot(upreg.count_longer,aes(x=temperature, y=count,shape=replicate))+
  geom_point()+
  facet_wrap(~gene_name+expression,scales = "free_y")


```





```{r}



library(ggplot2)
library(scales) # needed for oob parameter


# Coerce to a data frame
edgeR2ResDF <- as.data.frame(qlf)

# Examine this data frame
head(edgeR2ResDF)

# Set a boolean column for significance
edgeR2ResDF$significant <- ifelse(edgeR2ResDF$FDR < .05, "Significant",NA)
edgeR2ResDF$salmon_tax <- rownames(edgeR2ResDF)
edgeR2ResDF <- left_join(edgeR2ResDF, egg_nog_genename,by="salmon_tax")
edgeR2ResDF$delabel <- NA
edgeR2ResDF$delabel[!is.na(edgeR2ResDF$significant)] <- edgeR2ResDF$gene_name[!is.na(edgeR2ResDF$significant)]
edgeR2ResDF$delabel[!is.na(edgeR2ResDF$significant)]
# Plot the results similar to edgeR2
ggplot(edgeR2ResDF, aes(logCPM, logFC, colour=significant)) + 
  geom_point(size=1) + 
  scale_y_continuous(limits=c(-3, 3)) +
  geom_hline(yintercept = 0, colour="gray",size=1.5) + 
  labs(x="mean of normalized logCPM", y="log fold change")+
    geom_text_repel(aes(label=delabel),color="black")+
  theme_bw()
head(edgeR2ResDF)
```


# GO and KEGG annotation

###KEGG term
```{r}

library(KEGGREST)
head(egg_nog)

# select KEGG pathway column from eggnog output
egg_nog_kegg <-  egg_nog %>% dplyr::select(c(V1,V10)) %>% dplyr::rename("salmon_tax"=V1,"KEGG_terms" =V10)
# get DEG genes
edgeR2ResDF_DEG <- edgeR2ResDF %>% subset(FDR < 0.05)
# merge gene name edgeR result with eggnog kegg pathway
salmonkeggDF<- left_join(edgeR2ResDF_DEG,egg_nog_kegg,by="salmon_tax")
# get all KEGG terms for all genes
kegg.terms <- salmonkeggDF[!is.na(salmonkeggDF$KEGG_terms) & salmonkeggDF$KEGG_terms != "",]$KEGG_terms
# get unique KEGG terms for genes identified in samples
kegg.terms <- unique(unlist(sapply(kegg.terms,function(x){unlist(strsplit(x,split=","))})))

# only keep kegg terms with "ko" identifier, ignore kegg pathwat with "map" identifiers (overlapped)
kegg.terms.map <- kegg.terms[sapply(kegg.terms, function(x){
  grepl("ko",x)
})]


# remove the "path:" in front of every KEGG identifier
kegg_pathway <- getKEGGPathwayNames(species.KEGG="ko") %>% mutate(kegg_id=sapply(PathwayID, function(x){
  unlist(strsplit(x, split=":"))[2]
}))
kegg.term.list <- kegg_pathway$kegg_id

# make gene set for each KEGG identifier (all genes identified with the KEGG identifier will be put in a gene set)
index <- list()
for(term in kegg.terms.map){
  index_term.de <- which(unlist(lapply(salmonkeggDF$KEGG_terms, function(x){ifelse(grepl(term,x), TRUE, FALSE)}))) # get index of the kegg found in de.genes datafrane
  gene.sets <- salmonkeggDF[c(index_term.de),"salmon_tax"]# return the gene names with kegg term
  index_term <- as.numeric(unlist(ids2indices(gene.sets , rownames(y$counts), remove.empty=TRUE)))# find the index of these genes in the overall gene dataframe
  index[[eval(term)]] <- index_term # set into index list
}

rownames(y) %in% gene.sets
# only keep KEGG gene sets with more than 4 genes
kegg.pathway.diffexp <- camera(y, index,design=design) %>% subset(NGenes >=4)
# order differentially expressed KEGG path base on FDR
kegg.pathway.diffexp[order(kegg.pathway.diffexp$FDR),]
kegg.pathway.diffexp$kegg_id <- rownames(kegg.pathway.diffexp)
enriched.kegg <- left_join(kegg.pathway.diffexp,kegg_pathway)
enriched.kegg
# write.csv(enriched.kegg, "/Users/rachel/Dropbox/5.Rachel-projects/Transcriptomics_PolyATail/directRNA.Samples.Results/pathway.goterms.enrichment/kegg.pathway.enrichment.stats.csv",row.names = FALSE)

# left_join(kegg.pathway.diffexp,kegg_pathway) %>% select(-c(PValue.Mixed, FDR.Mixed, kegg_id))
# kegg_path_FDR_matrix <- head(left_join(kegg.pathway.diffexp,kegg_pathway) #%>% select(FDR,Description),5)
# rownames(kegg_path_FDR_matrix) <- kegg_path_FDR_matrix$Description
# kegg_path_FDR_matrix$FDR <- -log10(as.numeric(kegg_path_FDR_matrix$FDR))
# kegg_path_FDR_matrix <- as.matrix(kegg_path_FDR_matrix %>% select(!Description))
# pheatmap(kegg_path_FDR_matrix, cluster_cols=FALSE,cluster_rows = FALSE)
```
### explore genes in the enriched pathway 
```{r}

kegg.genes.enriched <- index$ko04212

y$counts[c(kegg.genes.enriched),]
```

###GO term
```{r}
library(GO.db)

head(egg_nog)
egg_nog_go <-  egg_nog %>% dplyr::select(c(V1,V7)) %>% dplyr::rename("salmon_tax"=V1,"GO_terms" =V7)
edgeR2ResDF_DEG <- edgeR2ResDF %>% subset(FDR < 0.05)
salmonGODF<- left_join(edgeR2ResDF_DEG,egg_nog_go,by="salmon_tax")
go.terms <- salmonGODF[!is.na(salmonGODF$GO_terms) & salmonGODF$GO_terms != "",]$GO_terms
go.terms <- unique(unlist(sapply(go.terms,function(x){unlist(strsplit(x,split=","))})))
length(go.terms)

zz = Ontology(GOTERM)
 table(unlist(zz))
 
 # get each go term's annotation
goterms = unlist(Term(GOTERM))
go.annotation <- data.frame(goterms)
go.annotation$GO_terms <- rownames(go.annotation)
 
 # database query all GO terms
MF.go <- Ontology(GOTERM)[Ontology(GOTERM)=="MF"]
BP.go <- Ontology(GOTERM)[Ontology(GOTERM)=="BP"]
CC.go <- Ontology(GOTERM)[Ontology(GOTERM)=="CC"]



# classify go terms analyzed from sequences based on queried GO terms
go.terms.mf <- go.terms[go.terms %in% names(MF.go)]
go.terms.bp <- go.terms[go.terms %in% names(BP.go)]
go.terms.cc <- go.terms[go.terms %in% names(CC.go)]




index <- list() # assign genes into each GO terms, each GO terms as a gene set
i=0
for(term in go.terms.bp){
  index_term <- which(unlist(lapply(salmonGODF$GO_terms, function(x){ifelse(grepl(term,x), TRUE, FALSE)})))
  index[[eval(term)]] <- index_term
  i=i+1
  # print(i)
}



# test the gene set (go term that is enriched), MF only 
go.diffexp <- camera(y, index,design=design) %>% subset(NGenes >=4)
go.diffexp$GO_terms <- rownames(go.diffexp)
go.diffexp <- go.diffexp[order(go.diffexp$FDR),]
go.diffexp.annot<-left_join(go.diffexp, go.annotation)
go.diffexp.annot
# go.diffexp.annot %>% dplyr::select(c(NGenes, Direction, PValue,FDR, GO_terms, goterms))

# write.csv(go.diffexp.annot, "/Users/rachel/Dropbox/5.Rachel-projects/Transcriptomics_PolyATail/directRNA.Samples.Results/pathway.goterms.enrichment/GO.Cellular.Components.Terms.enrichment.csv", row.names = FALSE)

# # if want to test for significance in specific pathway
# index <- list() # assign genes into each GO terms, each GO terms as a gene set
# i=0
# GO.interest <- c("GO:0045260", "GO:0043590")
# for(term in GO.interest){
#   index_term <- which(unlist(lapply(salmonGODF$GO_terms, function(x){ifelse(grepl(term,x), TRUE, FALSE)})))
#   index[[eval(term)]] <- index_term
#   i=i+1
#   # print(i)
# }
# 
# # test the gene set (go term that is enriched), MF only 
# go.diffexp <- fry(y, index,design=design) %>% subset(NGenes >=4)
# go.diffexp$GO_terms <- rownames(go.diffexp)
# go.diffexp <- go.diffexp[order(go.diffexp$FDR),]
# go.diffexp.annot<-left_join(go.diffexp, go.annotation)
# go.diffexp.annot %>% dplyr::select(c(NGenes, Direction, PValue,FDR, GO_terms, goterms))




```

### explore genes in the enriched GO Terms
```{r}

bp.genes.enriched.GO0017148 <- index$`GO:0017148`
salmonGODF[c(bp.genes.enriched.GO0017148),]

bp.genes.enriched.GO0032269 <- index$`GO:0032269`
salmonGODF[c(bp.genes.enriched.GO0032269),]

```