---
title: "stringtie_quan_diffExp"
author: "Rachel Xu"
date: "4/8/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(tximport)
library(ggplot2)
library(dplyr)
library(tidyr)
library(genefilter)
library(limma)
library(edgeR)
library(IsoformSwitchAnalyzeR)

user <- "rachel"

data_dir <- paste0(paste0("/Users/",user),"/Dropbox/5.Rachel-projects/Transcriptomics_PolyATail/quant_results/stringtie/direct_RNA/")
dir <-paste0(paste0("/Users/",user),"/Dropbox/5.Rachel-projects/Transcriptomics_PolyATail/quant_results/")
ref <- paste0(paste0("/Users/",user),"/Dropbox/5.Rachel-projects/Transcriptomics_PolyATail/quant_results/reference/")
```

# PART I. Isoform Switching Analysis

### 1) read in stringtie output
```{r}

# Read in stringtie results
stringtie_quant <- importIsoformExpression(
    parentDir = data_dir, readLength = 10000
)

head(stringtie_quant$abundance,2)


```
### read in sample and condition metadata 
```{r}

# direct_flair is metadata for the four direct RNA samples, self-generated
samps <- read.csv(paste(dir, "direct_flair.txt", sep="/"), sep="\t")
samps$condition <- factor(samps$condition)
table(samps$condition)
samps <- samps %>% dplyr::select(-c(X)) %>% dplyr::rename( sampleID = full_name )
```

### build an IsoformSwitchAnalyze for sownstream analysis
```{r}
# import stringtie results into R for downstream analysis
switchList <- importRdata(
    isoformCountMatrix   = stringtie_quant$counts,
    isoformRepExpression = stringtie_quant$abundance,
    designMatrix         = samps, # sample annotations
    isoformExonAnnoation =paste0(data_dir,"direct_RNA_gff.annotated.gtf") ,# isoform annotations (stringtie output)
    showProgress = FALSE
)

head(switchList$exons,2)

```

### filter transcripts base on expression level and number of isoforms
```{r}

switchListFiltered <- preFilter(
    switchAnalyzeRlist = switchList,
    geneExpressionCutoff = 1,
    isoformExpressionCutoff =0,
    removeSingleIsoformGenes = TRUE
)


```

### Using DEXSeq to perform isoform switching test
```{r}

# Due to small number of replicates, less than 3, the isoform switc test would be not relevant, we set all q-values to 1, and set downstram qvalue pvalue cutoff to >1 
switchListFiltered$isoformFeatures$gene_switch_q_value <- 1
switchListAnalyzed <- isoformSwitchTestDEXSeq(
    switchAnalyzeRlist = switchListFiltered,
    reduceToSwitchingGenes=FALSE,
    alpha=1
)

extractSwitchSummary(switchListAnalyzed, alpha =1.1) # 
```

```{r}

gtf <- file.path(data_dir,"direct_RNA_stringtie.denov.merged.gtf")

addORFfromGTF(switchListFiltered, gtf)
analyzeNovelIsoformORF(switchListFiltered)

```



# PART II. Ballgown DTU
- code below derived from worlflow below: https://pydupont.github.io/BallGownTuto/Protocol.html
```{r}

library(ballgown)
metadata_stringtie <-read.csv(paste(dir, "direct_flair.txt", sep="/"), sep="\t")
metadata_stringtie <- metadata_stringtie[,-2]
metadata_stringtie

stringtie_samples<- c(
"Q29_Copenhageni_Basecalled-June_11_2020_Repeat_Direct-RNA",
"Q29_Copenhageni_Basecalled_May_22_2020_Direct-RNA",
"Q36_Copenhageni_Basecalled_June_9_2020-Repeat_Direct-RNA",
"Q36_Copenhageni_Basecalled_May_31_2020_Direct-RNA")

stringtie_folders <- file.path(dir,"stringtie","direct_RNA", stringtie_samples)
ballgown_data <- ballgown(samples = stringtie_folders,
pData=metadata_stringtie)

ballgown_data_fil = subset(ballgown_data,"rowVars(texpr(ballgown_data)) >
1",genomesubset=TRUE)

# 
# ballgown_transcript <- 
# data.frame(geneNames=ballgown::geneNames(ballgown_data_fil),
# geneIDs=ballgown::geneIDs(ballgown_data_fil), ballgown_transcript)
# 
# 
# results_transcripts <- arrange(ballgown_transcript,pval)
# results_genes <- arrange(ballgown_gene,pval)

```
### visualize ballgown analysis
```{r}


fpkm<- texpr(ballgown_data,meas="all")
fpkm
fpkmDF = as.data.frame(fpkm)
fpkmDF = log2(fpkmDF+1)# to avoid 0
head(fpkmDF,2)
colnames(fpkmDF) <- c("Q29.Copenhageni.2","Q29.Copenhageni.1","Q36.Copenhageni.2","Q36.Copenhageni.1")

fpkm_longer <- as.data.frame(fpkmDF) %>% pivot_longer(c("Q29.Copenhageni.2","Q29.Copenhageni.1","Q36.Copenhageni.2","Q36.Copenhageni.1"),names_to = "samples", values_to = "Log2fold FPKM") %>% mutate(Temperature=sapply(samples,function(x){
  unlist(strsplit(x,".",fixed = TRUE))[1]
}))

ggplot(fpkm_longer,aes(x=samples,y=`Log2fold FPKM`))+
  geom_violin(aes(fill=Temperature))+
  geom_boxplot(width=0.1)+
  theme_bw()



```


```{r}

plotTranscripts(ballgown::geneIDs(ballgown_data)[5], ballgown_data,
main=c("Gene XIST in sample "), sample=c("Q36_Copenhageni_Basecalled_June_9_2020-Repeat_Direct-RNA"))

```


##  DESEQ
http://www.bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html

### results obtained using prepDE.py3
```{r}
library(DESeq2)
counts <- read.csv(paste0(data_dir,"gene_count_matrix.csv"), row.names = 1)

condition <- as.factor(metadata_stringtie$condition )

sample_names <- as.vector(metadata_stringtie$full_name )

coldata <- data.frame(row.names=sample_names, temperature = condition)

ballgown_data.deseq <- DESeqDataSetFromMatrix(counts, coldata, design=~ temperature)


deseq <- DESeq(ballgown_data.deseq)
deseq.results <- DESeq2::results(deseq)
deseq.results<-deseq.results[order(deseq.results$padj),]

# length(deseq.results[deseq.results$pvalue <0.05])
# deseq.results$pvalue <0.05
head(deseq.results,10)
```
### shrunken log fold change (LFC) to reduce the effect of dispersion
```{r}
resultsNames(deseq)
res_lfcShrink <- lfcShrink(deseq,coef = 2, type="apeglm")
res_lfcShrink<-res_lfcShrink[order(deseq.results$padj),]
sum(res_lfcShrink$padj < 0.05 , na.rm=TRUE) # 479 genes with differential expression > 0.05
```

```{r}
# read DESeq results in to a data frame
deseq_stringtieDF <- as.data.frame(res_lfcShrink)

# add a column for differential expression based on p value and log fold change
deseq_stringtieDF$diffexpressed <- "Not differentially expressed"

# if with p-value < 0.05 and log2fold change > abs(1), label either up or down regulated 
deseq_stringtieDF$diffexpressed[deseq_stringtieDF$log2FoldChange > 1 & deseq_stringtieDF$pvalue < 0.05] <- "Upregulated"
deseq_stringtieDF$diffexpressed[deseq_stringtieDF$log2FoldChange < -1 & deseq_stringtieDF$pvalue < 0.05] <- "Downregulated"

# factor so legend is ordered
deseq_stringtieDF$diffexpressed <- factor(deseq_stringtieDF$diffexpressed, levels=c("Upregulated", "Downregulated", "Not differentially expressed"))
head(deseq_stringtieDF)

# all transcript with associated gene names
deseq_stringtieDF$gene_symbol <- sapply(rownames(deseq_stringtieDF), function(x){
  unlist(strsplit(x, "|", fixed = TRUE))[2]
})
deseq_stringtieDF$delabel <- NA
deseq_stringtieDF$delabel[deseq_stringtieDF$diffexpressed != "Not differentially expressed"] <- deseq_stringtieDF$gene_symbol[deseq_stringtieDF$diffexpressed != "Not differentially expressed"]


library(ggrepel)
ggplot(deseq_stringtieDF, aes(x=log2FoldChange,y = -log10(padj)))+
  geom_point(aes(color =diffexpressed))+
  theme_bw()+
  geom_vline(xintercept=c(-1,1), col="gray",linetype="dashed")+
  geom_hline(yintercept=-log10(0.05), col="gray",linetype="dashed")+
  scale_color_manual(values=c("orange", "blue", "black"))+
  geom_text_repel(aes(label=delabel))
```
```{r}

# plotMA(res_lfcShrink)

library(ggplot2)
library(scales) # needed for oob parameter
library(viridis)

# Coerce to a data frame
deseq2ResDF <- as.data.frame(res_lfcShrink)

# Examine this data frame
head(deseq2ResDF)

# Set a boolean column for significance
deseq2ResDF$significant <- ifelse(deseq2ResDF$padj < .05 & (deseq2ResDF$log2FoldChange > 1 | deseq2ResDF$log2FoldChange < -1), "Significant",NA)
deseq2ResDF$gene_symbol <- sapply(rownames(deseq2ResDF), function(x){
  unlist(strsplit(x, "|", fixed = TRUE))[2]
})
deseq2ResDF$delabel <- NA
deseq2ResDF$delabel[!is.na(deseq2ResDF$significant)] <- deseq2ResDF$gene_symbol[!is.na(deseq2ResDF$significant)]
deseq2ResDF

# Plot the results similar to DEseq2
ggplot(deseq2ResDF, aes(baseMean, log2FoldChange, colour=significant)) + 
  geom_point(size=1) + 
  scale_y_continuous(limits=c(-3, 3)) +
  scale_x_log10() + 
  geom_hline(yintercept = 0, colour="gray",size=1.5) + 
  geom_vline(xintercept=100, col="gray",linetype="dashed")+
  labs(x="mean of normalized counts", y="log fold change")+
  geom_text_repel(aes(label=delabel),color="black")+
  theme_bw()


```
### heatmap
```{r}

# transforms the count data to the log2 scale in a way which minimizes differences between samples for rows with small counts, and which normalizes with respect to library size.
rld <- rlog(ballgown_data.deseq, blind=FALSE) # input is stringite output

# assay is a summarizedExperiment object, matrix like
topVarGenes <- head(order(rowVars(assay(rld)), decreasing=T),50)# top 100 genes' variances 
matrix <- assay(rld)[ topVarGenes, ] # extract genes with top variation aceoss samples
matrix <- matrix - rowMeans(matrix)
colnames(matrix)<- c("Q29.2","Q29.1","Q36.2","Q36.1")
# select the 'contrast' you want
annotation_data <- data.frame("samples"=c(rep("Q29",2), rep("Q36",2)))
rownames(annotation_data) <- colnames(matrix)
annoCol <- list(category = c("Q29","Q36"))

pheatmap(matrix, annotation_col = annotation_data, annotation_colors =annoCol, angle_col=45,fontsize_row=5)
```

### EdgeR

### results obtained using prepDE.py3
```{r}
library(edgeR)
counts <- read.csv(paste0(data_dir,"gene_count_matrix.csv"), row.names = 1)
group <- factor(c("Q29","Q29","Q36","Q36"))
y <- DGEList(counts=counts,group=group)
keep <- filterByExpr(y) # keeps rows that have worthwhile counts in a minumum numberof samples
y <- y[keep,,keep.lib.sizes=FALSE]
y <- calcNormFactors(y)
design <- model.matrix(~group)
y <- estimateDisp(y,design)
fit <- glmQLFit(y,design)
qlf <- glmQLFTest(fit,coef=2)
topTags(qlf,1000)

summary(qlf)
qlf$table$FDR <- p.adjust(qlf$table$PValue, method="BH")
head(qlf$table)

```

```{r}

# read DESeq results in to a data frame
edgeR_stringtieDF <- as.data.frame(qlf)
# write.csv(res_lfcShrink_stringtie,paste0(dir,"deseq_stringtieDF.csv"))
# add a column for differential expression based on p value and log fold change
edgeR_stringtieDF$diffexpressed <- "Not differentially expressed"

# if with p-value < 0.05 and log2fold change > abs(1), label either up or down regulated 
edgeR_stringtieDF$diffexpressed[edgeR_stringtieDF$logFC >= 1 & edgeR_stringtieDF$FDR < 0.05] <- "Upregulated"
edgeR_stringtieDF$diffexpressed[edgeR_stringtieDF$logFC <= -1 & edgeR_stringtieDF$FDR < 0.05] <- "Downregulated"
edgeR_stringtieDF
# factor so legend is ordered
edgeR_stringtieDF$diffexpressed <- factor(edgeR_stringtieDF$diffexpressed, levels=c("Upregulated", "Downregulated", "Not differentially expressed"))
edgeR_stringtieDF

# all transcript with associated gene names
edgeR_stringtieDF$gene_symbol <- sapply(rownames(edgeR_stringtieDF), function(x){
  unlist(strsplit(x, "|", fixed = TRUE))[2]
})

edgeR_stringtieDF$delabel <- NA
edgeR_stringtieDF$delabel[edgeR_stringtieDF$diffexpressed != "Not differentially expressed"] <- edgeR_stringtieDF$gene_symbol[edgeR_stringtieDF$diffexpressed != "Not differentially expressed"]
edgeR_stringtieDF
edgeR_stringtieDF[!is.na(edgeR_stringtieDF$delabel),]
library(ggrepel)
ggplot(edgeR_stringtieDF, aes(x=logFC,y = -log10(FDR)))+
  geom_point(aes(color =diffexpressed))+
  theme_bw()+
  geom_vline(xintercept=c(-1,1), col="gray",linetype="dashed")+
  geom_hline(yintercept=-log10(0.05), col="gray",linetype="dashed")+
  scale_color_manual(values=c("orange", "blue", "black"))+
  geom_text_repel(aes(label=delabel))

```

 


