---
title: "tailfindr_analysis"
author: "Rachel Xu"
date: "8/3/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}

library(dplyr)
```


```{r}
tailfindr.path <- "/Users/rx32940/Dropbox/5.Rachel-projects/Transcriptomics_PolyATail/polyATail_cDNA_Copen_072021/"


transcript.annot <- read.csv(file.path(tailfindr.path, "reference","GCF_000007685.1_ASM768v1_feature_table.txt"), sep="\t") %>% subset(X..feature == "CDS") %>% select(-c("X..feature", "class", "assembly", "assembly_unit", "seq_type", "related_accession", "GeneID", "locus_tag","attributes"))

```


## Summary statistics
```{r, warning=FALSE, message=FALSE}

SAMPLES <- list.dirs(file.path(tailfindr.path, "tailfindr"), recursive = FALSE)
possible_polyas <- c("noPolyA", "PolyA", "noPolyA_Q")

combined_tailfindr_stat <- data.frame(read_type=character(), tail_is_valid=logical(), read_type_num = integer(), SampleID = character(), Map_To = character(), Sample=character(), polya_status=character())

combined_tailLen <- data.frame(read_type=character(), tail_is_valid=logical(), tail_length=integer(), SampleID = character(), Map_To = character(), Sample=character(), polya_status=character())

  for (dir in SAMPLES){
  sample = basename(dir)
  POLYAS <- list.dirs(dir, recursive = FALSE) 
  POLYAS_DIRS <- POLYAS[ grepl("PolyA", basename(POLYAS))]
  
  for (polya in POLYAS_DIRS){
    
    
    tailfindr_out <- list.files(polya, recursive = FALSE)
    tailfindr_out <- tailfindr_out[grepl("annotated",tailfindr_out)]
    
    for (out in tailfindr_out){
    

    if(grepl("trans", out)){
      map_to_type <- "transcriptome"
      file.name <- paste0(sample, "_", basename(polya), ".annotated.trans.csv")
      sign <- "t"
    }else{
      map_to_type <- "genome"
      file.name <- paste0(sample, "_", basename(polya), ".annotated.csv")
      sign <- "g"
    }
      
      
    ###################summary statistics#####################################################################
      sample_abbr_name <- paste0(substring(sample,1,5), "_",basename(polya) )
      
      assign(eval(sample_abbr_name),read.csv(file.path(polya,file.name)) )
      assign(eval(paste0(sample_abbr_name, "_anot","_",sign)),get(sample_abbr_name) %>% group_by(read_type,tail_is_valid) %>%summarise(read_type_num=n()))
      assign(eval(paste0(sample_abbr_name, "_anot","_",sign)),get(paste0(sample_abbr_name, "_anot","_",sign)) %>% mutate(SampleID = sample_abbr_name))
      assign(eval(paste0(sample_abbr_name, "_anot","_",sign)),get(paste0(sample_abbr_name, "_anot","_",sign)) %>% mutate(Map_To = map_to_type))
      assign(eval(paste0(sample_abbr_name, "_anot","_",sign)),get(paste0(sample_abbr_name, "_anot","_",sign)) %>% mutate(Sample = sample))
      assign(eval(paste0(sample_abbr_name, "_anot","_",sign)),get(paste0(sample_abbr_name, "_anot","_",sign)) %>% mutate(polya_status = basename(polya)))
      
      combined_tailfindr_stat <- rbind( combined_tailfindr_stat, get(paste0(sample_abbr_name, "_anot","_",sign)))
    ####################tail length statistics#################################################################  
      sample_tail_abbr <- paste0(substring(sample,1,5), "_",basename(polya), "_tailLen","_",sign )
      
      assign(eval(sample_tail_abbr),get(sample_abbr_name)%>% subset(!is.na(tail_length)) %>% select(c(read_type, tail_is_valid, tail_length)))
      assign(eval(sample_tail_abbr),get(sample_tail_abbr) %>% mutate(SampleID = sample_abbr_name))
      assign(eval(sample_tail_abbr),get(sample_tail_abbr) %>% mutate(MAP_To = map_to_type))
      assign(eval(sample_tail_abbr),get(sample_tail_abbr) %>% mutate(Sample = sample))
      assign(eval(sample_tail_abbr),get(sample_tail_abbr) %>% mutate(polya_status= basename(polya)))
      
      combined_tailLen <- rbind(combined_tailLen, get(sample_tail_abbr))
    
    }
  }
  
  }

Icter_noPolyA_anot_t

combined_tailLen


```


### MAP to Transcriptome
```{r}

combined_transcript_mapped_top10 <- data.frame(product_accession = character(), num_reads_mapped = integer(), read_id=character(), read_type = character(), tail_is_valid = logical(),tail_start=integer(), tail_end=integer(), tail_length=integer(), chromsome=character(), start=integer(), end=integer(), non.redundant_refseq = character(), name=character(), symbol=character(), feature_interval_length=integer(), product_length=integer(), SampleID=character(), Sample=character(), polya_status = character())
SAMPLES <- list.dirs(file.path(tailfindr.path, "tailfindr"), recursive = FALSE)
for (dir in SAMPLES){
  
  sample = basename(dir)
  POLYAS <- list.dirs(dir, recursive = FALSE) 
  POLYAS_DIRS <- POLYAS[ grepl("PolyA", basename(POLYAS))]
  
    for (polya in POLYAS_DIRS){
      print(polya)
      tailfindr_out <- list.files(polya, recursive = FALSE)
      tailfindr_out <- tailfindr_out[grepl("annotated.trans.csv",tailfindr_out)]
      polya_base=basename(polya)
      
        for (out in tailfindr_out){
          print(out)
                copenNoPolyA_t <- read.csv(file.path(polya,out))
            copenNoPolyA_acc <-  copenNoPolyA_t %>%
              select(-c(file_path)) %>% mutate(product_accession=sapply(transcript_id, function(x){
                paste0("WP_",unlist(strsplit(x,"_", fixed = TRUE))[5])
              })) %>% mutate(genomic_accession = sapply(transcript_id, function(x){
                paste0("NC_",unlist(strsplit(x,"_", fixed = TRUE))[2])
              })) %>% select(-c(transcript_id))


            copenNoPolyA_annot_t <- left_join(copenNoPolyA_acc, transcript.annot, by=c("product_accession", "genomic_accession"))

            # write.csv(copenNoPolyA_annot_t,file.path(tailfindr.path, "tailfindr",sample,polya_base,paste0(sample,"_",polya_base,".trans.final.csv")), quote=FALSE, row.names=FALSE)
        
            # length(unique(copenNoPolyA_annot_t$non.redundant_refseq))
            # summary(copenNoPolyA_annot_t$non.redundant_refseq)
    
            transcript_mapped <- copenNoPolyA_annot_t %>% group_by(non.redundant_refseq, name,symbol, chromosome) %>% summarise(Num.Reads.Mapped=n()) %>% arrange(desc(Num.Reads.Mapped))
            
            # write.csv(transcript_mapped,file.path(tailfindr.path, "tailfindr",sample,polya_base,paste0(sample,"_",polya_base,".transcriptsMapped.stats.csv")), quote=FALSE, row.names=FALSE)
            
            # Number of Reads Mapped to each transcript
            transcript_mapped_count_table <- table(copenNoPolyA_annot_t$non.redundant_refseq)[order(table(copenNoPolyA_annot_t$non.redundant_refseq),decreasing =TRUE)]
            
            transcript_mapped_count_table_df <- data.frame(product_accession=names(transcript_mapped_count_table), num_reads_mapped=as.numeric(transcript_mapped_count_table))
            top_10_mapped_transcripts <- head(transcript_mapped_count_table_df$product_accession,10)
            
            transcript_mapped_count_table_df_annot <- left_join(transcript_mapped_count_table_df, copenNoPolyA_annot_t)
            
            transcript_mapped_transcripts_top10<- transcript_mapped_count_table_df_annot[transcript_mapped_count_table_df_annot$product_accession %in% top_10_mapped_transcripts,]
            
            sample_name <- paste0(substring(sample,1,5), "_", basename(polya))
            transcript_mapped_transcripts_top10$SampleID <- sample_name
            transcript_mapped_transcripts_top10$Sample <- sample
            transcript_mapped_transcripts_top10$polya_status <- basename(polya)
            transcript_mapped_transcripts_top10 <- transcript_mapped_transcripts_top10[c("product_accession", "num_reads_mapped", "read_id", "read_type", "tail_is_valid", "tail_is_valid", "tail_end", "tail_length", "chromosome", "start", "end", "non.redundant_refseq", "name",  "symbol", "feature_interval_length", "product_length", "SampleID", "Sample", "polya_status")]
            combined_transcript_mapped_top10 <- rbind(combined_transcript_mapped_top10 , transcript_mapped_transcripts_top10)

        }
    
      }
}

# 
# table(combined_transcript_mapped_top10$read_id)
# length(unique(copenNoPolyA_t$read_id))
```

# Combined

# Plot Read type summary
```{r}

combined_tailfindr_stat
library(ggplot2)
library(scales)
plot_map <- ggplot(combined_tailfindr_stat,aes(x=SampleID, y=read_type_num, fill=read_type))+
  geom_bar(stat = "identity", position = "stack")+
  facet_wrap( Sample~Map_To, scale="free",ncol = 2)+
  scale_y_continuous(label=comma)+
  theme_bw()+
  theme(axis.text = element_text(size=12),axis.text.x = element_text(angle = 45, vjust=0.99,hjust = 1), axis.title = element_text(size=12),strip.text = element_text(size=12),legend.text = element_text(size=12),legend.title = element_text(size=12))+
  labs(x="Samples", y= "Count of Read Types")
plot_map
ggsave(file.path(tailfindr.path,"tailfindr","figures" ,"Num_Read_Type_MapTo_Gen_Transcriptome.png"),plot_map,height = 10)

# write.csv(copenNoPolyA_annot, file.path(tailfindr.path, "tailfindr","noPolyA", "copenhageni_noPolyA.final.trans.csv"))
```


```{r}

combined_tailLen %>% subset(tail_is_valid==FALSE)
native_polya_mrna <- combined_tailLen %>% subset(Sample == "Copenhageni") %>% subset(grepl("noPolyA", polya_status))  %>% subset(MAP_To == "transcriptome")
native_polya_mrna
summary(native_polya_mrna$tail_length)

plot_tailLen <- ggplot(combined_tailLen,aes(x=SampleID, y=tail_length, fill = Sample ))+
      geom_violin()+
  facet_grid(read_type~MAP_To, scales = "free_y")+
  theme_bw()+
  theme(axis.text = element_text(size=12),axis.text.x = element_text(angle = 45, vjust=0.99,hjust = 1), axis.title = element_text(size=12),strip.text = element_text(size=12),legend.text = element_text(size=12),legend.title = element_text(size=12))+
  labs(y="Tail Length")
plot_tailLen
# ggsave(file.path(tailfindr.path,"tailfindr","figures" ,"tailLength_violin.pdf"),plot_tailLen,width = 8)
```

```{r}
library(tidyr)
library(ggplot2)

combined_transcript_mapped_top10
combined_transcript_mapped_top10<-combined_transcript_mapped_top10 %>% unite("x_label", c(product_accession,name), remove = FALSE,sep="\n")
combined_transcript_mapped_top10$x_label <- factor(combined_transcript_mapped_top10$x_label, levels = sort(unique(combined_transcript_mapped_top10$x_label)))
unique(combined_transcript_mapped_top10$polya_status)
plot_topMap <- ggplot(combined_transcript_mapped_top10 %>% subset(polya_status !="noPolyA_Q"), aes(x=x_label,fill=tail_is_valid))+
  geom_bar(position =position_dodge(preserve = "single"), stat = "count")+
  facet_wrap(Sample ~ polya_status , scale="free",ncol=2)+
  theme_bw()+
  theme(axis.text.x = element_text(angle = 90,hjust = 1, vjust =-0.001), axis.text=element_text(size=10), axis.title=element_text(size=12), legend.text = element_text(size=12), legend.title = element_text(size=12))+
  labs(y="Number of Reads Mapped to Transcript") +
  labs(x= "Transcripts ID/Names")

ggsave(file.path(tailfindr.path, "tailfindr", "figures", "Top10MapTranscripts.pdf"), plot_topMap,width = 12, height=20)
 plot_topMap
# # 
# plot_read <-ggplot(combined_transcript_mapped_top10, aes(x=x_label,fill=tail_is_valid))+
#   geom_bar(position=position_dodge(preserve="single") )+
#   facet_wrap(SampleID ~ read_type, scale="free", ncol = 3)+
#   theme_bw()+
#   theme(axis.text.x = element_text(angle = 90,hjust=1), axis.text=element_text(size=8), axis.title=element_text(size=12), legend.text = element_text(size=12), legend.title = element_text(size=12))+
#   labs(y="Number of Reads Mapped to Transcript") +
#   labs(x= "Transcripts ID/Names")
# # 
# plot_read
# ggsave(file.path(tailfindr.path, "tailfindr", "figures", "Top10MapTranscriptsReadType.pdf"), plot_read,width = 20,height = 40)

```
# Direct RNA samples tailfindr analysis output
```{r}

library(dplyr)
library(ggplot2)

# transcript annotation
tailfindr.path <- "/Users/rx32940/Dropbox/5.Rachel-projects/Transcriptomics_PolyATail/polyATail_cDNA_Copen_072021/"
transcript.annot <- read.csv(file.path(tailfindr.path, "reference","GCF_000007685.1_ASM768v1_feature_table.txt"), sep="\t") %>% subset(X..feature == "CDS") %>% select(-c("X..feature", "class", "assembly", "assembly_unit", "seq_type", "related_accession", "GeneID", "locus_tag","attributes"))


direct.tail.path <- "/Users/rx32940/Dropbox/5.Rachel-projects/Transcriptomics_PolyATail/directRNA_trans_09012021/tailfindr"

top_10_mapped_transcript_all_sample <- data.frame(non.redundant_refseq = character(), number.reads.mapped= character(), chromosome= character(), start=integer(), end=integer(), strand= character(), name= character(), symbol= character(), feature_interval_length=integer(), product_length=integer(), plot.label= character(), sample= character())

top_10_mapped_transcript_tail_all <- data.frame(non.redundant_refseq = character(), number.reads.mapped= character(), chromosome= character(), start=integer(), end=integer(), strand= character(), name= character(), symbol= character(), feature_interval_length=integer(), product_length=integer(), plot.label= character(), sample= character(), read_id= character(), tail_start=integer(), tail_end=integer(), tail_length=integer())
for(file in list.files(direct.tail.path, pattern = "*.annotated.csv")){
  
sample_name_str <- unlist(strsplit(file, "_", fixed = TRUE))
sample_name <- paste(sample_name_str[1], sample_name_str[2],sample_name_str[4] , sep="_")
direct.tailfindr.file <- read.csv(file.path(direct.tail.path, file)) %>% mutate(non.redundant_refseq = sapply(transcript_id, function(x){
  s1 <- unlist(strsplit(x, "_", fixed=TRUE))
  paste(s1[4], s1[5],sep="_")
}))

direct.tailfindr.map.count <- direct.tailfindr.file  %>% group_by(non.redundant_refseq) %>% summarise(number.reads.mapped=n()) %>% arrange(desc(number.reads.mapped))

top_10_mapped_transcript <- left_join(direct.tailfindr.map.count, transcript.annot, by="non.redundant_refseq") %>% select(c(
non.redundant_refseq, number.reads.mapped, chromosome, start, end, strand, name, symbol, feature_interval_length, product_length))%>% arrange(desc(number.reads.mapped)) %>% head(10) %>% mutate(plot.label=paste(non.redundant_refseq, name, sep="\n")) %>% mutate(sample=sample_name)

top_10_mapped_transcript_all_sample <- rbind(top_10_mapped_transcript_all_sample, top_10_mapped_transcript)

top_10_mapped_transcript_tail <- left_join(top_10_mapped_transcript, direct.tailfindr.file, by="non.redundant_refseq") %>% mutate(plot.label=paste(non.redundant_refseq, name, sep="\n")) %>% mutate(sample=sample_name) %>% select(-c("file_path","transcript_id", "samples_per_nt","mapping_quality", "sam_flag")) 

top_10_mapped_transcript_tail_all <- rbind(top_10_mapped_transcript_tail_all, top_10_mapped_transcript_tail)

}

top_10_mapped_transcript_all_sample %>% subset(non.redundant_refseq=="WP_000389641.1")

p1 <- ggplot(top_10_mapped_transcript_all_sample, aes(x=plot.label, y= number.reads.mapped))+
  geom_col()+
  facet_wrap(~sample, scale="free")+
  theme_bw()+
  theme(axis.text.x = element_text(angle = 90,hjust=0.95,vjust=0.2))+
  labs(x="Top 10 Mapped Transcripts")

p2 <- ggplot(top_10_mapped_transcript_tail_all, aes(x=plot.label, y=tail_length))+
  geom_violin()+
  geom_boxplot(width=0.1, outlier.shape = NA)+
  facet_wrap(~sample, scale="free")+
  theme_bw()+
  theme(axis.text.x = element_text(angle = 90,hjust=0.95,vjust=0.2))+
  labs(x="Top 10 Mapped Transcripts")
p2
ggsave(file.path(direct.tail.path, "figures", "tailfindr.top.10.mapped.transcript.pdf"), p1,height = 10,width = 9)
ggsave(file.path(direct.tail.path, "figures", "tailfindr.top.10.mapped.transcript.tailLength.pdf"), p2,height = 10,width = 9)

```
# compare read length from nanopolish and tailfindr
```{r}

tailfindr.read.length <- data.frame(read_id=character(), tail_length=integer())
for(file in list.files(direct.tail.path, pattern = "*.annotated.csv")){
tailfindr.file <- read.csv(file.path(direct.tail.path, file)) %>% select(c(read_id, tail_length)) %>% subset(!is.na(tail_length))
tailfindr.read.length <- rbind(tailfindr.read.length, tailfindr.file)
}
file
read.csv(file.path(direct.tail.path, "Q29_Copenhageni_Direct-RNA_FAST5_Pass.annotated.csv"))
nano.path <- "/Users/rx32940/Dropbox/5.Rachel-projects/Transcriptomics_PolyATail/directRNA_trans_09012021/nanopolish"
nanopolish.read.length <- data.frame(readname=character(), polya_length=integer())
for(file in list.files(nano.path, pattern="*trans.polyA.tsv")){
nanopolish.file <- read.csv(file.path(nano.path, "Q29_Copenhageni_Direct-RNA_FAST5_Pass.nanopolish.trans.polyA.tsv"), sep="\t") %>% subset(qc_tag == "PASS") %>% select(c(readname, polya_length))
}
colnames(nanopolish.file) <- c("read_id", "tail_length.nano")



combined_software <- full_join(tailfindr.read.length,nanopolish.file,by="read_id" ) %>% subset(!is.na(tail_length.nano))%>% subset(!is.na(tail_length))

p3 <- ggplot(combined_software, aes(x= tail_length, y= tail_length.nano))+
  geom_point()+
  theme_bw()+
  labs(x="Tailfindr polyA tail estimation", y = "Nanopolish polyA tail estimation")
p3


```
