---
title: "tu_analysis"
author: "Rachel Xu"
date: "2/14/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(dplyr)
library(tidyr)
library(ggplot2)
tu_path <- "/Users/rx32940/Dropbox/5.Rachel-projects/Transcriptomics_PolyATail/tu_annotation/TSS"


```


# identify parameters for TSS search
- each reads from cDNA and dRNA alignments were clustered with window size 10 - 100 bps to search for the best window size for TSS clustering

```{r}

tss_window_cDNA <- read.csv(file.path(tu_path, "output","window", "cDNA_samples_window_search_TSS.tab"), sep="\t") %>% mutate(samples=
  case_when(
    sample == "Copenhageni_Basecalled_Aug_16_2019_Direct-cDNA_NoPolyATail_Qiagen_rna_filtered" ~ "LIC-cDNA-nonpolyA_Q",
    sample =="Copenhageni_Basecalled_Aug_16_2019_Direct-cDNA_NoPolyATail_rna_filtered" ~ "LIC-cDNA-nonpolyA",
    sample =="Copenhageni_Basecalled_Aug_16_2019_Direct-cDNA_PolyATail_rna_filtered"~"LIC-cDNA-polyA"
  )
)

ggplot(tss_window_cDNA, aes(x=cluster_window, y= num_tss))+
  geom_point()+
  facet_wrap(vars(samples), ncol = 1,scales = "free_y")+
    scale_x_continuous(breaks = seq(0,150,5))+
  theme_bw()+
  labs(x="Window Size", y = "number of TSS")

tss_window_dRNA <- read.csv(file.path(tu_path, "output","window", "dRNA_samples_window_search_TSS.tab"), sep="\t") %>% mutate(samples=
  case_when(
    sample =="LIC_NOPOLYA_rna_filtered"~ "LIC-dRNA-nonpolyA",
    sample =="LIC_POLYA_DRNA_CDNA_rna_filtered" ~ "LIC-dRNA-cDNA-polyA",
    sample =="LIC_POLYA_rna_filtered" ~ "LIC-dRNA-polyA",
    sample =="Q29_Copenhageni_Basecalled_May_22_2020_Direct-RNA_rna_filtered" ~ "Q29-dRNA-nonpolyA-R1",
    sample =="Q29_Copenhageni_Basecalled-June_11_2020_Repeat_Direct-RNA_rna_filtered" ~ "Q29-dRNA-nonpolyA-R2",
    sample =="Q36_Copenhageni_Basecalled_June_9_2020-Repeat_Direct-RNA_rna_filtered" ~ "Q36-dRNA-nonpolyA-R2",
   sample =="Q36_Copenhageni_Basecalled_May_31_2020_Direct-RNA_rna_filtered" ~ "Q36-dRNA-nonpolyA-R1")
)


ggplot(tss_window_dRNA, aes(x=cluster_window, y= num_tss))+
  geom_point()+
  facet_wrap(vars(samples), ncol = 1,scales = "free_y")+
    scale_x_continuous(breaks = seq(0,150,5))+
  theme_bw()+
  labs(x="Window Size", y = "number of TSS")
```


# script to combine samples of each sequencing platform into one dataframe
- already moved to sapelo2 as ```combine_samples.r```
```{r}
# tabs <-list.files(file.path(tu_path, "input"), full.names = TRUE)
# 
# raw_tss_dfs <- lapply(tabs, function(x){read.table(x, sep="\t", header = TRUE) %>% select(!end) %>% mutate(sample=basename(x))})
# 
# 
# raw_tss_combined <- Reduce(function(...){full_join(..., all=TRUE,by=c("chrom", "strand", "gene", "start"))},raw_tss_dfs)
# 
# 
# all_samples_combined_raw <- raw_tss_combined %>% pivot_longer(cols=starts_with("sample"), names_to="colname", values_to = "files") %>% mutate(samples = 
#  case_when(
#     files == "Copenhageni_Basecalled_Aug_16_2019_Direct-cDNA_NoPolyATail_Qiagen_rna_filtered.combined_tss.tab" ~ "LIC-cDNA-nonpolyA_Q",
#     files =="Copenhageni_Basecalled_Aug_16_2019_Direct-cDNA_NoPolyATail_rna_filtered.combined_tss.tab" ~ "LIC-cDNA-nonpolyA",
#     files =="Copenhageni_Basecalled_Aug_16_2019_Direct-cDNA_PolyATail_rna_filtered.combined_tss.tab"~"LIC-cDNA-polyA",
#     files =="LIC_NOPOLYA_rna_filtered.combined_tss.tab"~ "LIC-dRNA-nonpolyA",
#     files =="LIC_POLYA_DRNA_CDNA_rna_filtered.combined_tss.tab" ~ "LIC-dRNA-cDNA-polyA",
#     files =="LIC_POLYA_rna_filtered.combined_tss.tab" ~ "LIC-dRNA-polyA",
#     files =="Q29_Copenhageni_Basecalled_May_22_2020_Direct-RNA_rna_filtered.combined_tss.tab" ~ "Q29-dRNA-nonpolyA-R1",
#     files =="Q29_Copenhageni_Basecalled-June_11_2020_Repeat_Direct-RNA_rna_filtered.combined_tss.tab" ~ "Q29-dRNA-nonpolyA-R2",
#     files =="Q36_Copenhageni_Basecalled_June_9_2020-Repeat_Direct-RNA_rna_filtered.combined_tss.tab" ~ "Q36-dRNA-nonpolyA-R2",
#    files =="Q36_Copenhageni_Basecalled_May_31_2020_Direct-RNA_rna_filtered.combined_tss.tab" ~ "Q36-dRNA-nonpolyA-R1")) %>%
#   select(-c("colname", "files")) %>% mutate(present = "x")  %>% subset(!is.na(samples))%>% pivot_wider(names_from = samples, values_from = present) %>% arrange(start)
# 
# chromIleadrawcDNA <- all_samples_combined_raw %>% subset(chrom != "NC_005824.1" & strand == "+") %>% select(!starts_with("Q36")) %>% select(c("chrom","start", "strand", "gene", starts_with("LIC-cDNA"))) %>% subset(!(is.na(`LIC-cDNA-nonpolyA_Q`) & is.na(`LIC-cDNA-nonpolyA`) & is.na(`LIC-cDNA-polyA`)))
# chromIlagrawcDNA <- all_samples_combined_raw %>% subset(chrom != "NC_005824.1" & strand == "-") %>% select(!starts_with("Q36")) %>% select(c("chrom","start", "strand", "gene", starts_with("LIC-cDNA")))%>% subset(!(is.na(`LIC-cDNA-nonpolyA_Q`) & is.na(`LIC-cDNA-nonpolyA`) & is.na(`LIC-cDNA-polyA`)))
# chromIIleadrawcDNA <- all_samples_combined_raw %>% subset(chrom != "NC_005823.1" & strand == "+") %>% select(!starts_with("Q36")) %>% select(c("chrom","start", "strand", "gene", starts_with("LIC-cDNA")))%>% subset(!(is.na(`LIC-cDNA-nonpolyA_Q`) & is.na(`LIC-cDNA-nonpolyA`) & is.na(`LIC-cDNA-polyA`)))
# chromIIlagrawcDNA <- all_samples_combined_raw %>% subset(chrom != "NC_005823.1" & strand == "-") %>% select(!starts_with("Q36")) %>% select(c("chrom","start", "strand", "gene", starts_with("LIC-cDNA")))%>% subset(!(is.na(`LIC-cDNA-nonpolyA_Q`) & is.na(`LIC-cDNA-nonpolyA`) & is.na(`LIC-cDNA-polyA`)))
# 
# chromIleadrawdRNA <- all_samples_combined_raw %>% subset(chrom != "NC_005824.1" & strand == "+") %>% select(!starts_with("Q36")) %>% select(c("chrom","start", "strand", "gene", !starts_with("LIC-cDNA"))) %>% subset(!(is.na(`Q29-dRNA-nonpolyA-R1`) & is.na(`LIC-dRNA-nonpolyA`) & is.na(`Q29-dRNA-nonpolyA-R2`) & is.na(`LIC-dRNA-polyA`)))
# chromIlagrawdRNA <- all_samples_combined_raw %>% subset(chrom != "NC_005824.1" & strand == "-") %>% select(!starts_with("Q36")) %>% select(c("chrom","start", "strand", "gene", !starts_with("LIC-cDNA")))%>% subset(!(is.na(`Q29-dRNA-nonpolyA-R1`) & is.na(`LIC-dRNA-nonpolyA`) & is.na(`Q29-dRNA-nonpolyA-R2`) & is.na(`LIC-dRNA-polyA`)))
# chromIIleadrawdRNA <- all_samples_combined_raw %>% subset(chrom != "NC_005823.1" & strand == "+") %>% select(!starts_with("Q36")) %>% select(c("chrom","start", "strand", "gene", !starts_with("LIC-cDNA")))%>% subset(!(is.na(`Q29-dRNA-nonpolyA-R1`) & is.na(`LIC-dRNA-nonpolyA`) & is.na(`Q29-dRNA-nonpolyA-R2`) & is.na(`LIC-dRNA-polyA`)))
# chromIIlagrawdRNA <- all_samples_combined_raw %>% subset(chrom != "NC_005823.1" & strand == "-") %>% select(!starts_with("Q36")) %>% select(c("chrom","start", "strand", "gene", !starts_with("LIC-cDNA")))%>% subset(!(is.na(`Q29-dRNA-nonpolyA-R1`) & is.na(`LIC-dRNA-nonpolyA`) & is.na(`Q29-dRNA-nonpolyA-R2`) & is.na(`LIC-dRNA-polyA`)))
# 
# chrom_strand_df <- c("chromIleadrawcDNA", "chromIlagrawcDNA", "chromIIleadrawcDNA", "chromIIlagrawcDNA", "chromIleadrawdRNA", "chromIlagrawdRNA", "chromIIleadrawdRNA", "chromIIlagrawdRNA")
# 
# for(df in chrom_strand_df){
#   print(get(df))
#   write.csv(get(df), file.path(tu_path,"output","raw_TSS" ,paste0(df, "_raw_TSS.csv")), sep=",", quote = FALSE)
#   
# }
# 
# all_samples_combined_raw%>% subset(start=="2109140")
```
# cluster TSS
- TSS obtained from different samples were clustered group_by chrom and strand
- ```cluster_tss.py``` script were used to obtain clusters of TSS
  - the script takes the TSS annotated from the same chrom and strand from the block above
  - add cluster ID as a new column
  
### TSS overlap between seqencing method replicates
```{r}
####################################################################################################
#### cDNA check 
########################################################################################################
# all_cdna_sample_window <- data.frame(window_size=as.integer(), l1vl2=as.integer(),l2vl3=as.integer(),l1vl3=as.integer(),intersect=as.integer())
# for (window_size in seq(0,300,10)){
window_size <- 10
lead_TSS_files_cDNA <- list.files(file.path(tu_path, "output", "raw_TSS_clustered"), pattern = paste0("*cDNA_raw_TSS_clustered_", window_size,".csv") ,full.names = TRUE)

cDNA_lead_tss <- lapply(lead_TSS_files_cDNA, function(x){read.csv(x, col.names = c("chrom",	"start",	"strand",	"gene",	"LIC.cDNA.nonpolyA_Q",	"LIC.cDNA.polyA",	"LIC.cDNA.nonpolyA", "cluster"))})

cDNA_lead_tss_df <- do.call(rbind, cDNA_lead_tss) %>% arrange(chrom)

l1 <- cDNA_lead_tss_df %>% subset(LIC.cDNA.nonpolyA_Q == "x")  %>% select(cluster) %>% unlist()
l2 <- cDNA_lead_tss_df %>% subset(LIC.cDNA.polyA == "x")  %>% select(cluster) %>% unlist()
l3 <- cDNA_lead_tss_df %>% subset(LIC.cDNA.nonpolyA == "x")  %>% select(cluster) %>% unlist()

# cur_window <- data.frame(window_size=window_size, l1vl2=length(intersect(l1,l2)), l2vl3=length(intersect(l2,l3)), l1vl3=length(intersect(l1,l3)), intersect=length(Reduce(intersect, list(l1,l2,l3))))
# all_cdna_sample_window <- rbind(all_cdna_sample_window,cur_window)
# }
# 
# all_cdna_sample_window_longer <- all_cdna_sample_window %>% pivot_longer(names_to = "comparison", values_to = "num_intersect", cols = starts_with("l"))
# 
# ggplot(all_cdna_sample_window_longer, aes(x=window_size, y=intersect))+
#   geom_point()+
#     scale_x_continuous(breaks = seq(0,300,10))+
#   theme_bw()

library(VennDiagram)
library(RColorBrewer)
library(UpSetR)


venn.diagram(
        x = list(l1, l2,l3),
        category.names = c("LIC.cDNA.nonpolyA_Q" , "LIC.cDNA.polyA","LIC.cDNA.nonpolyA"),
        filename = file.path(tu_path, "figures",'cDNA_replicates_TSS_intersect.png'),
        output=TRUE,
        fill=brewer.pal(3, "Pastel2"))

####################################################################################################
#### dRNA check 
########################################################################################################

# all_drna_sample_window <- data.frame(window_size=as.integer(), intersect=as.integer())
# for (window_size in seq(0,300,10)){

lead_TSS_files_drna <- list.files(file.path(tu_path, "output", "raw_TSS_clustered"), pattern = paste0("*rawdRNA_raw_TSS_clustered_",window_size,".csv") ,full.names = TRUE)

window_size <- 10

drna_lead_tss <- lapply(lead_TSS_files_drna, function(x){read.csv(x, col.names = c("chrom",	"start",	"strand",	"gene",	"LIC-dRNA-nonpolyA","Q29-dRNA-nonpolyA-R2","Q29-dRNA-nonpolyA-R1",		"LIC-dRNA-polyA",	"LIC-dRNA-cDNA-polyA", "cluster"))})

drna_lead_tss_df <- do.call(rbind, drna_lead_tss) %>% group_by(chrom, strand)%>%arrange(start, .by_group = TRUE)

l1 <- drna_lead_tss_df %>% subset(`LIC.dRNA.polyA` == "x")  %>% select(cluster) %>% unlist()
l2 <- drna_lead_tss_df %>% subset(`LIC.dRNA.cDNA.polyA` == "x")  %>% select(cluster) %>% unlist()
l3 <- drna_lead_tss_df %>% subset(`LIC.dRNA.nonpolyA` == "x")  %>% select(cluster) %>% unlist()
l4 <- drna_lead_tss_df %>% subset(`Q29.dRNA.nonpolyA.R1` == "x")  %>% select(cluster) %>% unlist()
l5 <- drna_lead_tss_df %>% subset(`Q29.dRNA.nonpolyA.R2` == "x")  %>% select(cluster) %>% unlist()
length(l5)
Reduce(intersect, list(l1,l3,l5))

# cur_window <- data.frame(window_size=window_size,  intersect=length(Reduce(intersect, list(l1,l2,l3,l4,l5))))
# all_drna_sample_window <- rbind(all_drna_sample_window,cur_window)
# }
# 
# # all_drna_sample_window_longer <- all_drna_sample_window %>% pivot_longer(names_to = "comparison", values_to = "num_intersect", cols = starts_with("l"))
# ggplot(all_drna_sample_window, aes(x=window_size, y=intersect))+
#   geom_point()+
#     scale_x_continuous(breaks = seq(0,300,10))+
#   theme_bw()
#   

library(VennDiagram)
library(RColorBrewer)
library(UpSetR)


venn.diagram(
        x = list(l1, l2,l3,l4,l5),
        category.names = c("LIC.dRNA.polyA" , "LIC.dRNA.cDNA.polyA","LIC.dRNA.nonpolyA", "Q29.dRNA.nonpolyA.R1", "Q29.dRNA.nonpolyA.R2"),
        filename = file.path(tu_path, "figures",'dRNA_replicates_TSS_intersect.png'),
        output=TRUE,
        fill=brewer.pal(5, "Pastel2"))

upset(fromList(list(LIC.dRNA.polyA=l1, LIC.dRNA.cDNA.polyA=l2,LIC.dRNA.nonpolyA=l3,Q29.dRNA.nonpolyA.R1=l4,Q29.dRNA.nonpolyA.R2=l5)), sets= c("LIC.dRNA.polyA" , "LIC.dRNA.cDNA.polyA","LIC.dRNA.nonpolyA", "Q29.dRNA.nonpolyA.R1", "Q29.dRNA.nonpolyA.R2"))  

```
### function to filter TSS for cDNA
```{r}



samples_cDNA <- c("LIC-cDNA-nonpolyA_Q",	"LIC-cDNA-polyA","LIC-cDNA-nonpolyA")
samples_dRNA <- c("LIC-dRNA-nonpolyA","Q29-dRNA-nonpolyA-R1","Q29-dRNA-nonpolyA-R2",		"LIC-dRNA-cDNA-polyA",	"LIC-dRNA-polyA")

#######################################################################
#
# filter TSS for cDNA sequenced samples
#
#######################################################################
# file_name <- lead_TSS_files_cDNA[2]
filter_lead_TSS_cDNA <- function(file_name){
chromI_lead_tss <- read.csv(file_name, row.names = 1, col.names = c("chrom",	"start",	"strand",	"gene",	samples_cDNA , "cluster")) 

# # TSS cluster with support of at least two samples were selected 
# cluster_filtered <- chromI_lead_tss %>% pivot_longer(cols = c(starts_with("LIC")), names_to = "samples", values_to = "present") %>% subset(!is.na(present)) %>% group_by(cluster) %>% mutate(sample_present=n_distinct(samples)) %>% subset(!sample_present < 2) %>% select(cluster) %>% unlist()

# in every cluster, the most far left (5') site were identified as teh TSS of the cluster, only cluster with at least 2 samples were selected
chromI_lead_tss_filtered <- chromI_lead_tss %>% pivot_longer(cols = c(starts_with("LIC")), names_to = "samples", values_to = "present") %>% subset(!is.na(present)) %>%  group_by(cluster) %>% mutate(TSS_min = min(start)) %>% select(-start) %>% ungroup() %>% distinct()%>% pivot_wider(names_from = samples, values_from = present) #%>% subset(!(is.na(LIC.cDNA.polyA) & !cluster %in% cluster_filtered)) %>% distinct( .keep_all = TRUE) # filter,drop TSS didn't find in the polyA sample and only existed in one of the nonpolyA samples


chromI_lead_tss_filtered
}

#######################################################################
#
# filter TSS for dRNA sequenced samples
#
#######################################################################
# file_name <- lead_TSS_files_dRNA[2]

filter_lead_TSS_dRNA <- function(file_name){
chromI_lead_tss <- read.csv(file_name, row.names = 1, col.names = c("chrom",	"start",	"strand",	"gene",	samples_dRNA , "cluster"))

# # TSS cluster with support of at least two samples were selected 
# cluster_filtered <- chromI_lead_tss %>% pivot_longer(cols = c(starts_with("LIC"), starts_with("Q")), names_to = "samples", values_to = "present") %>% subset(!is.na(present)) %>% group_by(cluster) %>% mutate(sample_present=n_distinct(samples))  %>% subset(!sample_present < 2) %>% select(cluster) %>% unlist()

# in every cluster, the most far left (5') site were identified as teh TSS of the cluster, only cluster with at least 2 samples were selected
chromI_lead_tss_filtered <- chromI_lead_tss %>% pivot_longer(cols = c(starts_with("LIC"), starts_with("Q")), names_to = "samples", values_to = "present") %>% subset(!is.na(present)) %>%  group_by(cluster) %>% mutate(TSS_min = min(start)) %>% select(-start) %>% ungroup() %>% distinct() %>% pivot_wider(names_from = samples, values_from = present)#%>% subset(!((is.na(LIC.dRNA.cDNA.polyA) | is.na(LIC.dRNA.polyA)) & !cluster %in% cluster_filtered)) %>% distinct(.keep_all = TRUE) # drop TSS didn't identify in either of the polyA dRNA samples, and also only existed once in other samples.


chromI_lead_tss_filtered
}
```

### combine TSS from cDNA and dRNA
```{r}

# identify TSS using cDNA samples
lead_TSS_files_cDNA <- list.files(file.path(tu_path, "output", "raw_TSS_clustered"), pattern = "*leadrawcDNA_raw_TSS_clustered_10.csv", full.names = TRUE)

lead_TSS_filtered_list <- lapply(lead_TSS_files_cDNA, filter_lead_TSS_cDNA)

cDNA_TSS_final <- do.call(rbind, lead_TSS_filtered_list)%>%group_by(chrom) %>%arrange(TSS_min,.by_group = TRUE) %>% distinct(cluster, .keep_all = TRUE)# %>% select(!c("cluster")) %>% distinct(start, .keep_all = TRUE) %>% arrange(start)


# identify TSS using dRNA samples
lead_TSS_files_dRNA <- list.files(file.path(tu_path, "output", "raw_TSS_clustered"), pattern = "*leadrawdRNA_raw_TSS_clustered_10.csv", full.names = TRUE)

lead_TSS_filtered_list <- lapply(lead_TSS_files_dRNA, filter_lead_TSS_dRNA)

dRNA_TSS_final <-  do.call(rbind, lead_TSS_filtered_list) %>%group_by(chrom) %>%arrange(TSS_min,.by_group = TRUE) %>% distinct(cluster, .keep_all = TRUE)#%>% select(!c("cluster"))

cDNA_TSS_final$seq_method <- "cDNA"
dRNA_TSS_final$seq_method <- "dRNA"

dRNA_cDNA_lead_combined <- rbind(cDNA_TSS_final %>% select(!starts_with("LIC")), dRNA_TSS_final %>% select(!c(starts_with("LIC"), starts_with("Q")))) %>% group_by(chrom) %>% arrange(TSS_min, .by_group = TRUE) %>% select(!cluster)

# write.csv(dRNA_cDNA_lead_combined, file.path(tu_path,"output","combined", "cDNA_dRNA_lead_TSS.csv"), quote=FALSE)
```

### filter cDNA and dRNA combined TSS
- use ```cluster_tss.py``` script on sapelo2 to cluster TSS annotated from cDNA and dRNA
```{r}
# samples_all <- c("LIC-cDNA-nonpolyA_Q",	"LIC-cDNA-nonpolyA",	"LIC-cDNA-polyA",	"Q29-dRNA-nonpolyA-R1",	"LIC-dRNA-nonpolyA","Q29-dRNA-nonpolyA-R2",	"LIC-dRNA-cDNA-polyA",	"LIC-dRNA-polyA")
# 
# cDNA_dRNA_lead_TSS_filtered <- read.csv(file.path(tu_path,"output","combined", "cDNA_dRNA_lead_TSS_clustered_100.csv"), col.names = c("chrom",		"strand",	"gene","TSS", "seq_method","cluster")) %>% subset(TSS !=1)%>% group_by(cluster) %>% mutate(TSS_min = min(TSS)) %>% group_by(gene, cluster) %>% distinct(TSS_min, .keep_all = TRUE)
# 
# # write.csv(cDNA_dRNA_lead_TSS_filtered , file.path(tu_path,"output","combined", "cDNA_dRNA_lead_TSS_filtered_nondistinct.csv"), quote=FALSE)
# 
# cDNA_dRNA_lead_TSS_distinct <- cDNA_dRNA_lead_TSS_filtered %>% ungroup() %>% distinct(TSS_min, .keep_all = TRUE)
# # write.csv(cDNA_dRNA_lead_TSS_filtered , file.path(tu_path,"output","combined", "cDNA_dRNA_lead_TSS_filtered_distinct.csv"), quote=FALSE)
```

### relationship between cDNA identified TSS vs dRNA identified TSS
- gene aware
```{r}


### intersecting TSS from cDNA and dRNA samples at different window size

clustered_TSS_files <- list.files(file.path(tu_path,"output", "combined"), pattern = "cDNA_dRNA_lead_TSS_clustered_*", full.names = TRUE)
intersect_size_windows <- data.frame(windowsize=as.integer(),TSS_intersect_size=as.integer())

for(cfile in clustered_TSS_files){

cDNA_dRNA_TSS <- read.csv(cfile, row.names = 1, col.names = c("X","chrom","strand",	"gene","TSS", "seq_method","cluster"))
s <- unlist(strsplit(basename(cfile), "[_.]"))
window_size <- as.integer(s[length(s)-1])

l1 <- cDNA_dRNA_TSS %>% subset(seq_method == "cDNA") %>% select(cluster) %>% unlist()
l2 <- cDNA_dRNA_TSS %>% subset(seq_method == "dRNA") %>% select(cluster) %>% unlist()

TSS_intersect_size <- length(intersect(l1, l2))

cur_window <- data.frame(window_size = window_size, TSS_intersect_size = TSS_intersect_size)
intersect_size_windows <- rbind(intersect_size_windows,cur_window)
}

# after window size of 30, the number of intersection between cDNA and dRNA samples begun to converge, we will use 30 as the window size 
library(ggplot2)
ggplot(intersect_size_windows, aes(x=window_size, y= TSS_intersect_size))+
  geom_point()+
  theme_bw()+
  scale_x_continuous(breaks = seq(0,150,10))+
  scale_y_continuous(breaks = seq(0,10000,100))



###### after elow plot, window size of 30 was chosen as the window size to converge 


cDNA_dRNA_TSS <- read.csv(file.path(tu_path,"output", "combined","cDNA_dRNA_lead_TSS_clustered_110.csv"), row.names = 1, col.names = c("chrom",		"strand",	"gene","TSS", "seq_method","cluster"))

l1 <- cDNA_dRNA_TSS %>% subset(seq_method == "cDNA") %>% select(cluster) %>% unlist()
l2 <- cDNA_dRNA_TSS %>% subset(seq_method == "dRNA") %>% select(cluster) %>% unlist()

library(VennDiagram)
library(RColorBrewer)

venn.diagram(
        x = list(l1, l2),
        category.names = c("cDNA" , "dRNA"),
        filename = file.path(tu_path, "figures",'cDNA_dRNA_TSS_intersect.png'),
        output=TRUE,
        fill=brewer.pal(2, "Pastel2")[1:2])


```
### what are the relationships between cDNA and dRNA samples' TSS in the same cluster
```{r}

# intersect_TSS_cluster <- cDNA_dRNA_TSS %>% subset(cluster %in% intersect(l1, l2)) %>% select(!gene) %>% distinct()
# 
# 
# ######################################################################
# #
# # This loop is used for identify TSS from each seq_method with more than 1 TSS in the same cluster
# # if more than 1 cluster identified, then the second 1 will add _2 to the seq_method_2 column
# #
# ######################################################################
# tss_freq_cdna <- 1 # tss freq for cdna samples in a cluster
# tss_freq_drna <- 0 # tss freq for drna samples in a cluster
# seq_method_list <- c("cDNA")
# cur_cluster <- intersect_TSS_cluster[1,"cluster"]
# for(row in 2:nrow(intersect_TSS_cluster)){
#   cur_cluster <- intersect_TSS_cluster[row,"cluster"]
#   if(cur_cluster == intersect_TSS_cluster[row-1,"cluster"]){
#   # count tss freq in each seq method's samples
#   if(intersect_TSS_cluster[row,"seq_method"] == "dRNA"){
#     tss_freq_drna = tss_freq_drna +1
#       # if more than 1 tss of one seq method'TSS was reported in cluster
#     if(tss_freq_drna > 1){
#       seq_method_list <- c(seq_method_list, paste0("dRNA_", tss_freq_drna))
#     }else{
#       seq_method_list <- c(seq_method_list, intersect_TSS_cluster[row,"seq_method"])
#     }
#     
#   }else{
#     tss_freq_cdna = tss_freq_cdna + 1
#       # if more than 1 tss of one seq method'TSS was reported in cluster
#     if(tss_freq_cdna > 1){
#       seq_method_list <- c(seq_method_list, paste0("cDNA_", tss_freq_cdna))
#     }else{
#       seq_method_list <- c(seq_method_list, intersect_TSS_cluster[row,"seq_method"])
#     }
#   }
#   }else{
#     cur_seq_method <- intersect_TSS_cluster[row,"seq_method"]
#     seq_method_list <- c(seq_method_list, cur_seq_method)
#     if(cur_seq_method == "dRNA"){
#     tss_freq_cdna <- 0 # tss freq for cdna samples in a cluster
#     tss_freq_drna <- 1
#     }else{
#     tss_freq_cdna <- 1 # tss freq for cdna samples in a cluster
#     tss_freq_drna <- 0
#     }
#   }
# 
# }
# 
# intersect_TSS_cluster$seq_method_2 <- seq_method_list
# 
# ######################################################################
# intersect_TSS_cluster %>% select(!seq_method) %>% pivot_wider(names_from = seq_method_2, values_from = TSS) %>% mutate(dRNAvscDNA=dRNA-cDNA)%>% mutate(dRNAvscDNA_2=dRNA_2-cDNA_2)  %>% subset(dRNAvscDNA < 0)
# 
# cDNA_dRNA_TSS <- read.csv(file.path(tu_path,"output", "combined","cDNA_dRNA_lead_TSS_clustered_100.csv"), row.names = 1, col.names = c("chrom",		"strand",	"gene","TSS", "seq_method","cluster"))
# cDNA_dRNA_TSS_2 <- read.csv(file.path(tu_path,"output", "combined","cDNA_dRNA_lead_TSS_clustered_10.csv"), row.names = 1, col.names = c("chrom",		"strand",	"gene","TSS", "seq_method","cluster"))
# 
# left_join(cDNA_dRNA_TSS, cDNA_dRNA_TSS_2, by =c("chrom", "strand", "gene", "TSS", "seq_method"))

```

```{r}

# cDNA_dRNA_TSS <- read.csv(file.path(tu_path,"output", "combined","cDNA_dRNA_lead_TSS_clustered_80.csv"), row.names = 1, col.names = c("chrom",		"strand",	"gene","TSS", "seq_method","cluster"))
# 
# cDNA_TSS_per_gene <- data.frame(table(cDNA_dRNA_TSS %>% subset(seq_method == "cDNA")%>% subset(!is.na(gene) & gene !="") %>% select(gene)))
# cDNA_1TSS_gene <- cDNA_TSS_per_gene %>% subset(Freq == 1) %>% select(Var1) %>% unlist()
# dRNA_TSS_per_gene <- data.frame(table(cDNA_dRNA_TSS %>% subset(seq_method == "dRNA")%>% subset(!is.na(gene) & gene !="") %>% select(gene)))
# dRNANA_1TSS_gene <- dRNA_TSS_per_gene %>% subset(Freq == 1) %>% select(Var1) %>% unlist()
# 
# intersect_cDNA_dRNA_gene <- intersect(cDNA_1TSS_gene,dRNANA_1TSS_gene)
# 
# cDNA_dRNA_TSS_gene <-cDNA_dRNA_TSS %>% subset(gene %in% intersect_cDNA_dRNA_gene) %>% select(!cluster) %>% pivot_wider(names_from = seq_method, values_from = TSS) %>% mutate(dRNAvscDNA=dRNA-cDNA)
```

### for lagging stand tss identification
```{r}

# file_name <- lag_TSS_files[2]

samples_cDNA <- c("LIC-cDNA-nonpolyA_Q",		"LIC-cDNA-polyA","LIC-cDNA-nonpolyA")
samples_dRNA <- c("LIC-dRNA-nonpolyA","Q29-dRNA-nonpolyA-R2","Q29-dRNA-nonpolyA-R1","LIC-dRNA-polyA",	"LIC-dRNA-cDNA-polyA")

#######################################################################
#
# filter TSS for cDNA sequenced samples
#
#######################################################################
# file_name <- lag_TSS_files_cDNA[2]
filter_lag_TSS_cDNA <- function(file_name){
chromI_lag_tss <- read.csv(file_name, row.names = 1, col.names = c("chrom",	"start",	"strand",	"gene",	samples_cDNA , "cluster"))

# TSS cluster with support of at least two samples were selected 
cluster_filtered <- chromI_lag_tss %>% pivot_longer(cols = c(starts_with("LIC")), names_to = "samples", values_to = "present") %>% subset(!is.na(present)) %>% group_by(cluster) %>% count() %>% subset(!n < 2) %>% select(cluster) %>% unlist()

# in every cluster, the most far left (5') site were identified as teh TSS of the cluster, only cluster with at least 2 samples were selected
chromI_lag_tss_filtered <- chromI_lag_tss %>% pivot_longer(cols = c(starts_with("LIC")), names_to = "samples", values_to = "present") %>% subset(!is.na(present)) %>%  group_by(cluster) %>% mutate(TSS_max = max(start)) %>% select(-start) %>% ungroup() %>% distinct()%>% pivot_wider(names_from = samples, values_from = present)# %>% subset(!(is.na(LIC.cDNA.polyA) & !cluster %in% cluster_filtered)) %>% distinct( .keep_all = TRUE) # filter,drop TSS didn't find in the polyA sample and only existed in one of the nonpolyA samples

chromI_lag_tss_filtered
}

#######################################################################
#
# filter TSS for dRNA sequenced samples
#
#######################################################################
# file_name <- lag_TSS_files_dRNA[2]

filter_lag_TSS_dRNA <- function(file_name){
chromI_lag_tss <- read.csv(file_name, row.names = 1, col.names = c("chrom",	"start",	"strand",	"gene",	samples_dRNA , "cluster"))

# TSS cluster with support of at least two samples were selected 
cluster_filtered <- chromI_lag_tss %>% pivot_longer(cols = c(starts_with("LIC"), starts_with("Q")), names_to = "samples", values_to = "present") %>% subset(!is.na(present)) %>% group_by(cluster) %>% count() %>% subset(!n < 2) %>% select(cluster) %>% unlist()

# in every cluster, the most far left (5') site were identified as teh TSS of the cluster, only cluster with at least 2 samples were selected
chromI_lag_tss_filtered <- chromI_lag_tss %>% pivot_longer(cols = c(starts_with("LIC"), starts_with("Q")), names_to = "samples", values_to = "present") %>% subset(!is.na(present)) %>%  group_by(cluster) %>% mutate(TSS_max = max(start)) %>% select(-start) %>% ungroup() %>% distinct() %>% pivot_wider(names_from = samples, values_from = present)#%>% subset(!((is.na(LIC.dRNA.cDNA.polyA) | is.na(LIC.dRNA.polyA)) & !cluster %in% cluster_filtered)) %>% distinct(.keep_all = TRUE) # drop TSS didn't identify in both of the polyA dRNA samples, and also only existed once in other samples.


chromI_lag_tss_filtered 
}
```

```{r}

# identify TSS using cDNA samples
lag_TSS_files_cDNA <- list.files(file.path(tu_path, "output", "raw_TSS_clustered"), pattern = "*lagrawcDNA_raw_TSS_clustered_10.csv", full.names = TRUE)

lag_TSS_filtered_list <- lapply(lag_TSS_files_cDNA, filter_lag_TSS_cDNA)

cDNA_TSS_final <- do.call(rbind, lag_TSS_filtered_list)%>% group_by(chrom)  %>% arrange(TSS_max, .by_group = TRUE)%>% distinct(cluster, .keep_all = TRUE)%>% select(!c("cluster"))


# identify TSS using dRNA samples
lag_TSS_files_dRNA <- list.files(file.path(tu_path, "output", "raw_TSS_clustered"), pattern = "*lagrawdRNA_raw_TSS_clustered_10.csv", full.names = TRUE)

lag_TSS_filtered_list <- lapply(lag_TSS_files_dRNA, filter_lag_TSS_dRNA)

dRNA_TSS_final <-  do.call(rbind, lag_TSS_filtered_list) %>% arrange(chrom) %>% arrange(TSS_max, .by_group = TRUE)%>% distinct(cluster, .keep_all = TRUE)%>% select(!c("cluster"))


cDNA_TSS_final$seq_method <- "cDNA"
dRNA_TSS_final$seq_method <- "dRNA"

dRNA_cDNA_lag_combined <- rbind(cDNA_TSS_final %>% select(!starts_with("LIC")), dRNA_TSS_final %>% select(!c(starts_with("LIC"), starts_with("Q")))) %>% group_by(chrom) %>% arrange(TSS_max, .by_group = TRUE)
# write.csv(dRNA_cDNA_lag_combined, file.path(tu_path,"output", "combined","cDNA_dRNA_lag_TSS.csv"))
```
